<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Roadmap" href="roadmap.html" /><link rel="prev" title="Citing PopPUNK" href="citing.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2023.09.10 -->
        <title>Reference documentation - PopPUNK 2.6.2 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">PopPUNK 2.6.2 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="_static/poppunk_v2.png" alt="Logo"/>
  </div>
  
  <span class="sidebar-brand-text">PopPUNK 2.6.2 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">PopPUNK documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="sketching.html">Sketching (<code class="docutils literal notranslate"><span class="pre">--create-db</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="qc.html">Data quality control (<code class="docutils literal notranslate"><span class="pre">--qc-db</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_fitting.html">Fitting new models (<code class="docutils literal notranslate"><span class="pre">--fit-model</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_distribution.html">Distributing PopPUNK models</a></li>
<li class="toctree-l1"><a class="reference internal" href="query_assignment.html">Query assignment (<code class="docutils literal notranslate"><span class="pre">poppunk_assign</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="visualisation.html">Creating visualisations</a></li>
<li class="toctree-l1"><a class="reference internal" href="mst.html">Minimum spanning trees</a></li>
<li class="toctree-l1"><a class="reference internal" href="subclustering.html">Subclustering with PopPIPE</a></li>
<li class="toctree-l1"><a class="reference internal" href="gpu.html">Using GPUs</a></li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Scripts</a></li>
<li class="toctree-l1"><a class="reference internal" href="poppunk_iterate.html">Iterative PopPUNK</a></li>
<li class="toctree-l1"><a class="reference internal" href="citing.html">Citing PopPUNK</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Reference documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="roadmap.html">Roadmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="miscellaneous.html">Miscellaneous</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="reference-documentation">
<h1>Reference documentation<a class="headerlink" href="#reference-documentation" title="Permalink to this heading">#</a></h1>
<p>Documentation for module functions (for developers)</p>
<section id="assign-py">
<h2>assign.py<a class="headerlink" href="#assign-py" title="Permalink to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">poppunk_assign</span></code> main function</p>
<span class="target" id="module-PopPUNK.assign"></span><dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.assign.assign_query">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.assign.</span></span><span class="sig-name descname"><span class="pre">assign_query</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbFuncs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qc_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update_db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_references</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distances</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">serial</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_fit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph_weights</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_preserved</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">previous_clustering</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">external_clustering</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">core</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accessory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gpu_sketch</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gpu_dist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gpu_graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">deviceid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_partial_query_graph</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/assign.html#assign_query"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.assign.assign_query" title="Permalink to this definition">#</a></dt>
<dd><p>Code for assign query mode for CLI</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.assign.assign_query_hdf5">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.assign.</span></span><span class="sig-name descname"><span class="pre">assign_query_hdf5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbFuncs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qNames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qc_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update_db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_references</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distances</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">serial</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_fit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph_weights</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_preserved</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">previous_clustering</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">external_clustering</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">core</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accessory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gpu_dist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gpu_graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_partial_query_graph</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/assign.html#assign_query_hdf5"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.assign.assign_query_hdf5" title="Permalink to this definition">#</a></dt>
<dd><p>Code for assign query mode taking hdf5 as input. Written as a separate function so it can be called
by web APIs</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.assign.main">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.assign.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/assign.html#main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.assign.main" title="Permalink to this definition">#</a></dt>
<dd><p>Main function. Parses cmd line args and runs in the specified mode.</p>
</dd></dl>

</section>
<section id="bgmm-py">
<h2>bgmm.py<a class="headerlink" href="#bgmm-py" title="Permalink to this heading">#</a></h2>
<p>Functions used to fit the mixture model to a database. Access using
<a class="reference internal" href="#PopPUNK.models.BGMMFit" title="PopPUNK.models.BGMMFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">BGMMFit</span></code></a>.</p>
<span class="target" id="module-PopPUNK.bgmm"></span><p>BGMM using sklearn</p>
<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.bgmm.findWithinLabel">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.bgmm.</span></span><span class="sig-name descname"><span class="pre">findWithinLabel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">means</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assignments</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/bgmm.html#findWithinLabel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.bgmm.findWithinLabel" title="Permalink to this definition">#</a></dt>
<dd><p>Identify within-strain links</p>
<p>Finds the component with mean closest to the origin and also akes sure
some samples are assigned to it (in the case of small weighted
components with a Dirichlet prior some components are unused)</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>means (numpy.array)</dt><dd><p>K x 2 array of mixture component means</p>
</dd>
<dt>assignments (numpy.array)</dt><dd><p>Sample cluster assignments</p>
</dd>
<dt>rank (int)</dt><dd><p>Which label to find, ordered by distance from origin. 0-indexed.</p>
<p>(default = 0)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>within_label (int)</dt><dd><p>The cluster label for the within-strain assignments</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.bgmm.fit2dMultiGaussian">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.bgmm.</span></span><span class="sig-name descname"><span class="pre">fit2dMultiGaussian</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpgmm_max_K</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/bgmm.html#fit2dMultiGaussian"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.bgmm.fit2dMultiGaussian" title="Permalink to this definition">#</a></dt>
<dd><p>Main function to fit BGMM model, called from <a class="reference internal" href="#PopPUNK.models.BGMMFit.fit" title="PopPUNK.models.BGMMFit.fit"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit()</span></code></a></p>
<p>Fits the mixture model specified, saves model parameters to a file, and assigns the samples to
a component. Write fit summary stats to STDERR.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (np.array)</dt><dd><p>n x 2 array of core and accessory distances for n samples.
This should be subsampled to 100000 samples.</p>
</dd>
<dt>dpgmm_max_K (int)</dt><dd><p>Maximum number of components to use with the EM fit.
(default = 2)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>dpgmm (sklearn.mixture.BayesianGaussianMixture)</dt><dd><p>Fitted bgmm model</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.bgmm.log_likelihood">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.bgmm.</span></span><span class="sig-name descname"><span class="pre">log_likelihood</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">means</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">covars</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/bgmm.html#log_likelihood"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.bgmm.log_likelihood" title="Permalink to this definition">#</a></dt>
<dd><p>modified sklearn GMM function predicting distribution membership</p>
<p>Returns the mixture LL for points X. Used by <code class="xref py py-func docutils literal notranslate"><span class="pre">assign_samples()</span></code> and
<a class="reference internal" href="#PopPUNK.plot.plot_contours" title="PopPUNK.plot.plot_contours"><code class="xref py py-func docutils literal notranslate"><span class="pre">plot_contours()</span></code></a></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>n x 2 array of core and accessory distances for n samples</p>
</dd>
<dt>weights (numpy.array)</dt><dd><p>Component weights from <a class="reference internal" href="#PopPUNK.bgmm.fit2dMultiGaussian" title="PopPUNK.bgmm.fit2dMultiGaussian"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit2dMultiGaussian()</span></code></a></p>
</dd>
<dt>means (numpy.array)</dt><dd><p>Component means from <a class="reference internal" href="#PopPUNK.bgmm.fit2dMultiGaussian" title="PopPUNK.bgmm.fit2dMultiGaussian"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit2dMultiGaussian()</span></code></a></p>
</dd>
<dt>covars (numpy.array)</dt><dd><p>Component covariances from <a class="reference internal" href="#PopPUNK.bgmm.fit2dMultiGaussian" title="PopPUNK.bgmm.fit2dMultiGaussian"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit2dMultiGaussian()</span></code></a></p>
</dd>
<dt>scale (numpy.array)</dt><dd><p>Scaling of core and accessory distances from <a class="reference internal" href="#PopPUNK.bgmm.fit2dMultiGaussian" title="PopPUNK.bgmm.fit2dMultiGaussian"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit2dMultiGaussian()</span></code></a></p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>logprob (numpy.array)</dt><dd><p>The log of the probabilities under the mixture model</p>
</dd>
<dt>lpr (numpy.array)</dt><dd><p>The components of the log probability from each mixture component</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.bgmm.log_multivariate_normal_density">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.bgmm.</span></span><span class="sig-name descname"><span class="pre">log_multivariate_normal_density</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">means</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">covars</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_covar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-07</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/bgmm.html#log_multivariate_normal_density"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.bgmm.log_multivariate_normal_density" title="Permalink to this definition">#</a></dt>
<dd><p>Log likelihood of multivariate normal density distribution</p>
<p>Used to calculate per component Gaussian likelihood in
<code class="xref py py-func docutils literal notranslate"><span class="pre">assign_samples()</span></code></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>n x 2 array of core and accessory distances for n samples</p>
</dd>
<dt>means (numpy.array)</dt><dd><p>Component means from <a class="reference internal" href="#PopPUNK.bgmm.fit2dMultiGaussian" title="PopPUNK.bgmm.fit2dMultiGaussian"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit2dMultiGaussian()</span></code></a></p>
</dd>
<dt>covars (numpy.array)</dt><dd><p>Component covariances from <a class="reference internal" href="#PopPUNK.bgmm.fit2dMultiGaussian" title="PopPUNK.bgmm.fit2dMultiGaussian"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit2dMultiGaussian()</span></code></a></p>
</dd>
<dt>min_covar (float)</dt><dd><p>Minimum covariance, added when Choleksy decomposition fails
due to too few observations (default = 1.e-7)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>log_prob (numpy.array)</dt><dd><p>An n-vector with the log-likelihoods for each sample being in
this component</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="dbscan-py">
<h2>dbscan.py<a class="headerlink" href="#dbscan-py" title="Permalink to this heading">#</a></h2>
<p>Functions used to fit DBSCAN to a database. Access using
<a class="reference internal" href="#PopPUNK.models.DBSCANFit" title="PopPUNK.models.DBSCANFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">DBSCANFit</span></code></a>.</p>
<span class="target" id="module-PopPUNK.dbscan"></span><p>DBSCAN using hdbscan</p>
<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.dbscan.evaluate_dbscan_clusters">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.dbscan.</span></span><span class="sig-name descname"><span class="pre">evaluate_dbscan_clusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/dbscan.html#evaluate_dbscan_clusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.dbscan.evaluate_dbscan_clusters" title="Permalink to this definition">#</a></dt>
<dd><p>Evaluate whether fitted dbscan model contains non-overlapping clusters</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>model (DBSCANFit)</dt><dd><p>Fitted model from <a class="reference internal" href="#PopPUNK.models.DBSCANFit.fit" title="PopPUNK.models.DBSCANFit.fit"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit()</span></code></a></p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>indistinct (bool)</dt><dd><p>Boolean indicating whether putative within- and
between-strain clusters of points overlap</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.dbscan.findBetweenLabel">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.dbscan.</span></span><span class="sig-name descname"><span class="pre">findBetweenLabel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assignments</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">within_cluster</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/dbscan.html#findBetweenLabel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.dbscan.findBetweenLabel" title="Permalink to this definition">#</a></dt>
<dd><p>Identify between-strain links from a DBSCAN model</p>
<p>Finds the component containing the largest number of between-strain
links, excluding the cluster identified as containing within-strain
links.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>assignments (numpy.array)</dt><dd><p>Sample cluster assignments</p>
</dd>
<dt>within_cluster (int)</dt><dd><p>Cluster ID assigned to within-strain assignments, from <a class="reference internal" href="#PopPUNK.bgmm.findWithinLabel" title="PopPUNK.bgmm.findWithinLabel"><code class="xref py py-func docutils literal notranslate"><span class="pre">findWithinLabel()</span></code></a></p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>between_cluster (int)</dt><dd><p>The cluster label for the between-strain assignments</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.dbscan.fitDbScan">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.dbscan.</span></span><span class="sig-name descname"><span class="pre">fitDbScan</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_cluster_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_out</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/dbscan.html#fitDbScan"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.dbscan.fitDbScan" title="Permalink to this definition">#</a></dt>
<dd><p>Function to fit DBSCAN model as an alternative to the Gaussian</p>
<p>Fits the DBSCAN model to the distances using hdbscan</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (np.array)</dt><dd><p>n x 2 array of core and accessory distances for n samples</p>
</dd>
<dt>min_samples (int)</dt><dd><p>Parameter for DBSCAN clustering ‘conservativeness’</p>
</dd>
<dt>min_cluster_size (int)</dt><dd><p>Minimum number of points in a cluster for HDBSCAN</p>
</dd>
<dt>cache_out (str)</dt><dd><p>Prefix for DBSCAN cache used for refitting</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>hdb (hdbscan.HDBSCAN)</dt><dd><p>Fitted HDBSCAN to subsampled data</p>
</dd>
<dt>labels (list)</dt><dd><p>Cluster assignments of each sample</p>
</dd>
<dt>n_clusters (int)</dt><dd><p>Number of clusters used</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-PopPUNK.mandrake">
<span id="mandrake-py"></span><h2>mandrake.py<a class="headerlink" href="#module-PopPUNK.mandrake" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.mandrake.generate_embedding">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.mandrake.</span></span><span class="sig-name descname"><span class="pre">generate_embedding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqLabels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accMat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perplexity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kNN</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxIter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/mandrake.html#generate_embedding"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.mandrake.generate_embedding" title="Permalink to this definition">#</a></dt>
<dd><p>Generate t-SNE projection using accessory distances</p>
<p>Writes a plot of t-SNE clustering of accessory distances (.dot)</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>seqLabels (list)</dt><dd><p>Processed names of sequences being analysed.</p>
</dd>
<dt>accMat (numpy.array)</dt><dd><p>n x n array of accessory distances for n samples.</p>
</dd>
<dt>perplexity (int)</dt><dd><p>Perplexity parameter passed to t-SNE</p>
</dd>
<dt>outPrefix (str)</dt><dd><p>Prefix for all generated output files, which will be placed in
<cite>outPrefix</cite> subdirectory</p>
</dd>
<dt>overwrite (bool)</dt><dd><p>Overwrite existing output if present
(default = False)</p>
</dd>
<dt>kNN (int)</dt><dd><p>Number of neigbours to use with SCE (cannot be &gt; n_samples)
(default = 50)</p>
</dd>
<dt>maxIter (int)</dt><dd><p>Number of iterations to run
(default = 1000000)</p>
</dd>
<dt>n_threads (int)</dt><dd><p>Number of CPU threads to use
(default = 1)</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use GPU libraries</p>
</dd>
<dt>device_id (int)</dt><dd><p>Device ID of GPU to be used
(default = 0)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>mandrake_filename (str)</dt><dd><p>Filename with .dot of embedding</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-PopPUNK.models">
<span id="models-py"></span><h2>models.py<a class="headerlink" href="#module-PopPUNK.models" title="Permalink to this heading">#</a></h2>
<p>Classes used for model fits</p>
<dl class="py class">
<dt class="sig sig-object py" id="PopPUNK.models.BGMMFit">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PopPUNK.models.</span></span><span class="sig-name descname"><span class="pre">BGMMFit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50000</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#BGMMFit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.BGMMFit" title="Permalink to this definition">#</a></dt>
<dd><p>Class for fits using the Gaussian mixture model. Inherits from <a class="reference internal" href="#PopPUNK.models.ClusterFit" title="PopPUNK.models.ClusterFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterFit</span></code></a>.</p>
<p>Must first run either <a class="reference internal" href="#PopPUNK.models.BGMMFit.fit" title="PopPUNK.models.BGMMFit.fit"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit()</span></code></a> or <a class="reference internal" href="#PopPUNK.models.BGMMFit.load" title="PopPUNK.models.BGMMFit.load"><code class="xref py py-func docutils literal notranslate"><span class="pre">load()</span></code></a> before calling
other functions</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>outPrefix (str)</dt><dd><p>The output prefix used for reading/writing</p>
</dd>
<dt>max_samples (int)</dt><dd><p>The number of subsamples to fit the model to</p>
<p>(default = 100000)</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.BGMMFit.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">progress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#BGMMFit.assign"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.BGMMFit.assign" title="Permalink to this definition">#</a></dt>
<dd><p>Assign the clustering of new samples using <code class="xref py py-func docutils literal notranslate"><span class="pre">assign_samples()</span></code></p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>X (numpy.array)</dt><dd><p>Core and accessory distances</p>
</dd>
<dt>values (bool)</dt><dd><p>Return the responsibilities of assignment rather than most likely cluster</p>
</dd>
<dt>num_processes (int)</dt><dd><p>Number of threads to use</p>
</dd>
<dt>progress (bool)</dt><dd><p>Show progress bar</p>
<p>[default = True]</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>y (numpy.array)</dt><dd><p>Cluster assignments or values by samples</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.BGMMFit.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_components</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#BGMMFit.fit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.BGMMFit.fit" title="Permalink to this definition">#</a></dt>
<dd><p>Extends <a class="reference internal" href="#PopPUNK.models.ClusterFit.fit" title="PopPUNK.models.ClusterFit.fit"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit()</span></code></a></p>
<p>Fits the BGMM and returns assignments by calling
<a class="reference internal" href="#PopPUNK.bgmm.fit2dMultiGaussian" title="PopPUNK.bgmm.fit2dMultiGaussian"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit2dMultiGaussian()</span></code></a>.</p>
<p>Fitted parameters are stored in the object.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>The core and accessory distances to cluster. Must be set if
preprocess is set.</p>
</dd>
<dt>max_components (int)</dt><dd><p>Maximum number of mixture components to use.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>y (numpy.array)</dt><dd><p>Cluster assignments of samples in X</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.BGMMFit.load">
<span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fit_npz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_obj</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#BGMMFit.load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.BGMMFit.load" title="Permalink to this definition">#</a></dt>
<dd><p>Load the model from disk. Called from <a class="reference internal" href="#PopPUNK.models.loadClusterFit" title="PopPUNK.models.loadClusterFit"><code class="xref py py-func docutils literal notranslate"><span class="pre">loadClusterFit()</span></code></a></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>fit_npz (dict)</dt><dd><p>Fit npz opened with <code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.load()</span></code></p>
</dd>
<dt>fit_obj (sklearn.mixture.BayesianGaussianMixture)</dt><dd><p>The saved fit object</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.BGMMFit.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#BGMMFit.plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.BGMMFit.plot" title="Permalink to this definition">#</a></dt>
<dd><p>Extends <a class="reference internal" href="#PopPUNK.models.ClusterFit.plot" title="PopPUNK.models.ClusterFit.plot"><code class="xref py py-func docutils literal notranslate"><span class="pre">plot()</span></code></a></p>
<p>Write a summary of the fit, and plot the results using
<a class="reference internal" href="#PopPUNK.plot.plot_results" title="PopPUNK.plot.plot_results"><code class="xref py py-func docutils literal notranslate"><span class="pre">PopPUNK.plot.plot_results()</span></code></a> and <a class="reference internal" href="#PopPUNK.plot.plot_contours" title="PopPUNK.plot.plot_contours"><code class="xref py py-func docutils literal notranslate"><span class="pre">PopPUNK.plot.plot_contours()</span></code></a></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>Core and accessory distances</p>
</dd>
<dt>y (numpy.array)</dt><dd><p>Cluster assignments from <a class="reference internal" href="#PopPUNK.models.BGMMFit.assign" title="PopPUNK.models.BGMMFit.assign"><code class="xref py py-func docutils literal notranslate"><span class="pre">assign()</span></code></a></p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.BGMMFit.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#BGMMFit.save"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.BGMMFit.save" title="Permalink to this definition">#</a></dt>
<dd><p>Save the model to disk, as an npz and pkl (using outPrefix).</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="PopPUNK.models.ClusterFit">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PopPUNK.models.</span></span><span class="sig-name descname"><span class="pre">ClusterFit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_dtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">numpy.float32</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#ClusterFit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.ClusterFit" title="Permalink to this definition">#</a></dt>
<dd><p>Parent class for all models used to cluster distances</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>outPrefix (str)</dt><dd><p>The output prefix used for reading/writing</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.ClusterFit.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#ClusterFit.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.ClusterFit.copy" title="Permalink to this definition">#</a></dt>
<dd><p>Copy the model to a new directory</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.ClusterFit.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#ClusterFit.fit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.ClusterFit.fit" title="Permalink to this definition">#</a></dt>
<dd><p>Initial steps for all fit functions.</p>
<p>Creates output directory. If preprocess is set then subsamples passed X</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>X (numpy.array)</dt><dd><p>The core and accessory distances to cluster. Must be set if
preprocess is set.</p>
<p>(default = None)</p>
</dd>
<dt>default_dtype (numpy dtype)</dt><dd><p>Type to use if no X provided</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.ClusterFit.no_scale">
<span class="sig-name descname"><span class="pre">no_scale</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#ClusterFit.no_scale"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.ClusterFit.no_scale" title="Permalink to this definition">#</a></dt>
<dd><p>Turn off scaling (useful for refine, where optimization
is done in the scaled space).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.ClusterFit.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#ClusterFit.plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.ClusterFit.plot" title="Permalink to this definition">#</a></dt>
<dd><p>Initial steps for all plot functions.</p>
<p>Ensures model has been fitted.</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>X (numpy.array)</dt><dd><p>The core and accessory distances to subsample.</p>
<p>(default = None)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="PopPUNK.models.DBSCANFit">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PopPUNK.models.</span></span><span class="sig-name descname"><span class="pre">DBSCANFit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100000</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#DBSCANFit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.DBSCANFit" title="Permalink to this definition">#</a></dt>
<dd><p>Class for fits using HDBSCAN. Inherits from <a class="reference internal" href="#PopPUNK.models.ClusterFit" title="PopPUNK.models.ClusterFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterFit</span></code></a>.</p>
<p>Must first run either <a class="reference internal" href="#PopPUNK.models.DBSCANFit.fit" title="PopPUNK.models.DBSCANFit.fit"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit()</span></code></a> or <a class="reference internal" href="#PopPUNK.models.DBSCANFit.load" title="PopPUNK.models.DBSCANFit.load"><code class="xref py py-func docutils literal notranslate"><span class="pre">load()</span></code></a> before calling
other functions</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>outPrefix (str)</dt><dd><p>The output prefix used for reading/writing</p>
</dd>
<dt>max_samples (int)</dt><dd><p>The number of subsamples to fit the model to</p>
<p>(default = 100000)</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.DBSCANFit.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">progress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#DBSCANFit.assign"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.DBSCANFit.assign" title="Permalink to this definition">#</a></dt>
<dd><p>Assign the clustering of new samples using <code class="xref py py-func docutils literal notranslate"><span class="pre">assign_samples_dbscan()</span></code></p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>X (numpy.array)</dt><dd><p>Core and accessory distances</p>
</dd>
<dt>no_scale (bool)</dt><dd><p>Do not scale X</p>
<p>[default = False]</p>
</dd>
<dt>progress (bool)</dt><dd><p>Show progress bar</p>
<p>[default = True]</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>y (numpy.array)</dt><dd><p>Cluster assignments by samples</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.DBSCANFit.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_num_clusters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_cluster_prop</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#DBSCANFit.fit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.DBSCANFit.fit" title="Permalink to this definition">#</a></dt>
<dd><p>Extends <a class="reference internal" href="#PopPUNK.models.ClusterFit.fit" title="PopPUNK.models.ClusterFit.fit"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit()</span></code></a></p>
<p>Fits the distances with HDBSCAN and returns assignments by calling
<a class="reference internal" href="#PopPUNK.dbscan.fitDbScan" title="PopPUNK.dbscan.fitDbScan"><code class="xref py py-func docutils literal notranslate"><span class="pre">fitDbScan()</span></code></a>.</p>
<p>Fitted parameters are stored in the object.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>The core and accessory distances to cluster. Must be set if
preprocess is set.</p>
</dd>
<dt>max_num_clusters (int)</dt><dd><p>Maximum number of clusters in DBSCAN fitting</p>
</dd>
<dt>min_cluster_prop (float)</dt><dd><p>Minimum proportion of points in a cluster in DBSCAN fitting</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>y (numpy.array)</dt><dd><p>Cluster assignments of samples in X</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.DBSCANFit.load">
<span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fit_npz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_obj</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#DBSCANFit.load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.DBSCANFit.load" title="Permalink to this definition">#</a></dt>
<dd><p>Load the model from disk. Called from <a class="reference internal" href="#PopPUNK.models.loadClusterFit" title="PopPUNK.models.loadClusterFit"><code class="xref py py-func docutils literal notranslate"><span class="pre">loadClusterFit()</span></code></a></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>fit_npz (dict)</dt><dd><p>Fit npz opened with <code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.load()</span></code></p>
</dd>
<dt>fit_obj (hdbscan.HDBSCAN)</dt><dd><p>The saved fit object</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.DBSCANFit.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#DBSCANFit.plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.DBSCANFit.plot" title="Permalink to this definition">#</a></dt>
<dd><p>Extends <a class="reference internal" href="#PopPUNK.models.ClusterFit.plot" title="PopPUNK.models.ClusterFit.plot"><code class="xref py py-func docutils literal notranslate"><span class="pre">plot()</span></code></a></p>
<p>Write a summary of the fit, and plot the results using
<a class="reference internal" href="#PopPUNK.plot.plot_dbscan_results" title="PopPUNK.plot.plot_dbscan_results"><code class="xref py py-func docutils literal notranslate"><span class="pre">PopPUNK.plot.plot_dbscan_results()</span></code></a></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>Core and accessory distances</p>
</dd>
<dt>y (numpy.array)</dt><dd><p>Cluster assignments from <a class="reference internal" href="#PopPUNK.models.BGMMFit.assign" title="PopPUNK.models.BGMMFit.assign"><code class="xref py py-func docutils literal notranslate"><span class="pre">assign()</span></code></a></p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.DBSCANFit.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#DBSCANFit.save"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.DBSCANFit.save" title="Permalink to this definition">#</a></dt>
<dd><p>Save the model to disk, as an npz and pkl (using outPrefix).</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="PopPUNK.models.LineageFit">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PopPUNK.models.</span></span><span class="sig-name descname"><span class="pre">LineageFit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_search_depth</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reciprocal_only</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count_unique_distances</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dist_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#LineageFit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.LineageFit" title="Permalink to this definition">#</a></dt>
<dd><p>Class for fits using the lineage assignment model. Inherits from <a class="reference internal" href="#PopPUNK.models.ClusterFit" title="PopPUNK.models.ClusterFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterFit</span></code></a>.</p>
<p>Must first run either <a class="reference internal" href="#PopPUNK.models.LineageFit.fit" title="PopPUNK.models.LineageFit.fit"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit()</span></code></a> or <a class="reference internal" href="#PopPUNK.models.LineageFit.load" title="PopPUNK.models.LineageFit.load"><code class="xref py py-func docutils literal notranslate"><span class="pre">load()</span></code></a> before calling
other functions</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>outPrefix (str)</dt><dd><p>The output prefix used for reading/writing</p>
</dd>
<dt>ranks (list)</dt><dd><p>The ranks used in the fit</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.LineageFit.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rank</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#LineageFit.assign"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.LineageFit.assign" title="Permalink to this definition">#</a></dt>
<dd><p>Get the edges for the network. A little different from other methods,
as it doesn’t go through the long form distance vector (as coo_matrix
is basically already in the correct gt format)</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>rank (int)</dt><dd><p>Rank to assign at</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>y (list of tuples)</dt><dd><p>Edges to include in network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.LineageFit.edge_weights">
<span class="sig-name descname"><span class="pre">edge_weights</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rank</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#LineageFit.edge_weights"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.LineageFit.edge_weights" title="Permalink to this definition">#</a></dt>
<dd><p>Get the distances for each edge returned by assign</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>rank (int)</dt><dd><p>Rank assigned at</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>weights (list)</dt><dd><p>Distance for each assignment</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.LineageFit.extend">
<span class="sig-name descname"><span class="pre">extend</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">qqDists</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qrDists</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#LineageFit.extend"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.LineageFit.extend" title="Permalink to this definition">#</a></dt>
<dd><p>Update the sparse distance matrix of nearest neighbours after querying</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>qqDists (numpy or cupy ndarray)</dt><dd><p>Two column array of query-query distances</p>
</dd>
<dt>qqDists (numpy or cupy ndarray)</dt><dd><p>Two column array of reference-query distances</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>y (list of tuples)</dt><dd><p>Edges to include in network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.LineageFit.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accessory</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#LineageFit.fit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.LineageFit.fit" title="Permalink to this definition">#</a></dt>
<dd><p>Extends <a class="reference internal" href="#PopPUNK.models.ClusterFit.fit" title="PopPUNK.models.ClusterFit.fit"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit()</span></code></a></p>
<p>Gets assignments by using nearest neigbours.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>The core and accessory distances to cluster. Must be set if
preprocess is set.</p>
</dd>
<dt>accessory (bool)</dt><dd><p>Use accessory rather than core distances</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>y (numpy.array)</dt><dd><p>Cluster assignments of samples in X</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.LineageFit.load">
<span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fit_npz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_obj</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#LineageFit.load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.LineageFit.load" title="Permalink to this definition">#</a></dt>
<dd><p>Load the model from disk. Called from <a class="reference internal" href="#PopPUNK.models.loadClusterFit" title="PopPUNK.models.loadClusterFit"><code class="xref py py-func docutils literal notranslate"><span class="pre">loadClusterFit()</span></code></a></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>fit_npz (dict)</dt><dd><p>Fit npz opened with <code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.load()</span></code></p>
</dd>
<dt>fit_obj (sklearn.mixture.BayesianGaussianMixture)</dt><dd><p>The saved fit object</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.LineageFit.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#LineageFit.plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.LineageFit.plot" title="Permalink to this definition">#</a></dt>
<dd><p>Extends <a class="reference internal" href="#PopPUNK.models.ClusterFit.plot" title="PopPUNK.models.ClusterFit.plot"><code class="xref py py-func docutils literal notranslate"><span class="pre">plot()</span></code></a></p>
<p>Write a summary of the fit, and plot the results using
<a class="reference internal" href="#PopPUNK.plot.plot_results" title="PopPUNK.plot.plot_results"><code class="xref py py-func docutils literal notranslate"><span class="pre">PopPUNK.plot.plot_results()</span></code></a> and <a class="reference internal" href="#PopPUNK.plot.plot_contours" title="PopPUNK.plot.plot_contours"><code class="xref py py-func docutils literal notranslate"><span class="pre">PopPUNK.plot.plot_contours()</span></code></a></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>Core and accessory distances</p>
</dd>
<dt>y (any)</dt><dd><p>Unused variable for compatibility with other
plotting functions</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.LineageFit.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#LineageFit.save"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.LineageFit.save" title="Permalink to this definition">#</a></dt>
<dd><p>Save the model to disk, as an npz and pkl (using outPrefix).</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="PopPUNK.models.NumpyShared">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PopPUNK.models.</span></span><span class="sig-name descname"><span class="pre">NumpyShared</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shape</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#PopPUNK.models.NumpyShared" title="Permalink to this definition">#</a></dt>
<dd><dl class="py attribute">
<dt class="sig sig-object py" id="PopPUNK.models.NumpyShared.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#PopPUNK.models.NumpyShared.dtype" title="Permalink to this definition">#</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="PopPUNK.models.NumpyShared.name">
<span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#PopPUNK.models.NumpyShared.name" title="Permalink to this definition">#</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="PopPUNK.models.NumpyShared.shape">
<span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#PopPUNK.models.NumpyShared.shape" title="Permalink to this definition">#</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="PopPUNK.models.RefineFit">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PopPUNK.models.</span></span><span class="sig-name descname"><span class="pre">RefineFit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#RefineFit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.RefineFit" title="Permalink to this definition">#</a></dt>
<dd><p>Class for fits using a triangular boundary and network properties. Inherits from <a class="reference internal" href="#PopPUNK.models.ClusterFit" title="PopPUNK.models.ClusterFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterFit</span></code></a>.</p>
<p>Must first run either <a class="reference internal" href="#PopPUNK.models.RefineFit.fit" title="PopPUNK.models.RefineFit.fit"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit()</span></code></a> or <a class="reference internal" href="#PopPUNK.models.RefineFit.load" title="PopPUNK.models.RefineFit.load"><code class="xref py py-func docutils literal notranslate"><span class="pre">load()</span></code></a> before calling
other functions</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>outPrefix (str)</dt><dd><p>The output prefix used for reading/writing</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.RefineFit.apply_threshold">
<span class="sig-name descname"><span class="pre">apply_threshold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#RefineFit.apply_threshold"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.RefineFit.apply_threshold" title="Permalink to this definition">#</a></dt>
<dd><p>Applies a boundary threshold, given by user. Does not run
optimisation.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>The core and accessory distances to cluster. Must be set if
preprocess is set.</p>
</dd>
<dt>threshold (float)</dt><dd><p>The value along the x-axis (core distance) at which to
draw the assignment boundary</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>y (numpy.array)</dt><dd><p>Cluster assignments of samples in X</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.RefineFit.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slope</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#RefineFit.assign"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.RefineFit.assign" title="Permalink to this definition">#</a></dt>
<dd><p>Assign the clustering of new samples</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>Core and accessory distances</p>
</dd>
<dt>slope (int)</dt><dd><p>Override self.slope. Default - use self.slope
Set to 0 for a vertical line, 1 for a horizontal line, or
2 to use a slope</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>y (numpy.array)</dt><dd><p>Cluster assignments by samples</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.RefineFit.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_move</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_move</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">startFile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indiv_refine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unconstrained</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multi_boundary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_local</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">betweenness_sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#RefineFit.fit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.RefineFit.fit" title="Permalink to this definition">#</a></dt>
<dd><p>Extends <a class="reference internal" href="#PopPUNK.models.ClusterFit.fit" title="PopPUNK.models.ClusterFit.fit"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit()</span></code></a></p>
<p>Fits the distances by optimising network score, by calling
<code class="xref py py-func docutils literal notranslate"><span class="pre">refineFit2D()</span></code>.</p>
<p>Fitted parameters are stored in the object.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>The core and accessory distances to cluster. Must be set if
preprocess is set.</p>
</dd>
<dt>sample_names (list)</dt><dd><p>Sample names in X (accessed by <a class="reference internal" href="#PopPUNK.utils.iterDistRows" title="PopPUNK.utils.iterDistRows"><code class="xref py py-func docutils literal notranslate"><span class="pre">iterDistRows()</span></code></a>)</p>
</dd>
<dt>model (ClusterFit)</dt><dd><p>The model fit to refine</p>
</dd>
<dt>max_move (float)</dt><dd><p>Maximum distance to move away from start point</p>
</dd>
<dt>min_move (float)</dt><dd><p>Minimum distance to move away from start point</p>
</dd>
<dt>startFile (str)</dt><dd><p>A file defining an initial fit, rather than one from <code class="docutils literal notranslate"><span class="pre">--fit-model</span></code>.
See documentation for format.
(default = None).</p>
</dd>
<dt>indiv_refine (str)</dt><dd><p>Run refinement for core or accessory distances separately
(default = None).</p>
</dd>
<dt>multi_boundary (int)</dt><dd><p>Produce cluster output at multiple boundary positions downward
from the optimum.
(default = 0).</p>
</dd>
<dt>unconstrained (bool)</dt><dd><p>If True, search in 2D and change the slope of the boundary</p>
</dd>
<dt>score_idx (int)</dt><dd><p>Index of score from <a class="reference internal" href="#PopPUNK.network.networkSummary" title="PopPUNK.network.networkSummary"><code class="xref py py-func docutils literal notranslate"><span class="pre">networkSummary()</span></code></a> to use
[default = 0]</p>
</dd>
<dt>no_local (bool)</dt><dd><p>Turn off the local optimisation step.
Quicker, but may be less well refined.</p>
</dd>
<dt>betweenness_sample (int)</dt><dd><p>Number of sequences per component used to estimate betweenness using
a GPU. Smaller numbers are faster but less precise [default = 100]</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use cugraph for graph analyses</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>y (numpy.array)</dt><dd><p>Cluster assignments of samples in X</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.RefineFit.load">
<span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fit_npz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_obj</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#RefineFit.load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.RefineFit.load" title="Permalink to this definition">#</a></dt>
<dd><p>Load the model from disk. Called from <a class="reference internal" href="#PopPUNK.models.loadClusterFit" title="PopPUNK.models.loadClusterFit"><code class="xref py py-func docutils literal notranslate"><span class="pre">loadClusterFit()</span></code></a></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>fit_npz (dict)</dt><dd><p>Fit npz opened with <code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.load()</span></code></p>
</dd>
<dt>fit_obj (None)</dt><dd><p>The saved fit object (not used)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.RefineFit.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#RefineFit.plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.RefineFit.plot" title="Permalink to this definition">#</a></dt>
<dd><p>Extends <a class="reference internal" href="#PopPUNK.models.ClusterFit.plot" title="PopPUNK.models.ClusterFit.plot"><code class="xref py py-func docutils literal notranslate"><span class="pre">plot()</span></code></a></p>
<p>Write a summary of the fit, and plot the results using
<a class="reference internal" href="#PopPUNK.plot.plot_refined_results" title="PopPUNK.plot.plot_refined_results"><code class="xref py py-func docutils literal notranslate"><span class="pre">PopPUNK.plot.plot_refined_results()</span></code></a></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>Core and accessory distances</p>
</dd>
<dt>y (numpy.array)</dt><dd><p>Assignments (unused)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PopPUNK.models.RefineFit.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#RefineFit.save"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.RefineFit.save" title="Permalink to this definition">#</a></dt>
<dd><p>Save the model to disk, as an npz and pkl (using outPrefix).</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.models.assign_samples">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.models.</span></span><span class="sig-name descname"><span class="pre">assign_samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chunk</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#assign_samples"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.assign_samples" title="Permalink to this definition">#</a></dt>
<dd><p>Runs a models assignment on a chunk of input</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>chunk (int)</dt><dd><p>Index of chunk to process</p>
</dd>
<dt>X (NumpyShared)</dt><dd><p>n x 2 array of core and accessory distances for n samples</p>
</dd>
<dt>y (NumpyShared)</dt><dd><p>An n-vector to store results, with the most likely cluster memberships
or an n by k matrix with the component responsibilities for each sample.</p>
</dd>
<dt>weights (numpy.array)</dt><dd><p>Component weights from <a class="reference internal" href="#PopPUNK.models.BGMMFit" title="PopPUNK.models.BGMMFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">BGMMFit</span></code></a></p>
</dd>
<dt>means (numpy.array)</dt><dd><p>Component means from <a class="reference internal" href="#PopPUNK.models.BGMMFit" title="PopPUNK.models.BGMMFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">BGMMFit</span></code></a></p>
</dd>
<dt>covars (numpy.array)</dt><dd><p>Component covariances from <a class="reference internal" href="#PopPUNK.models.BGMMFit" title="PopPUNK.models.BGMMFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">BGMMFit</span></code></a></p>
</dd>
<dt>scale (numpy.array)</dt><dd><p>Scaling of core and accessory distances from <a class="reference internal" href="#PopPUNK.models.BGMMFit" title="PopPUNK.models.BGMMFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">BGMMFit</span></code></a></p>
</dd>
<dt>chunk_size (int)</dt><dd><p>Size of each chunk in X</p>
</dd>
<dt>values (bool)</dt><dd><p>Whether to return the responsibilities, rather than the most
likely assignment (used for entropy calculation).</p>
<p>Default is False</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.models.loadClusterFit">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.models.</span></span><span class="sig-name descname"><span class="pre">loadClusterFit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pkl_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">npz_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/models.html#loadClusterFit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.models.loadClusterFit" title="Permalink to this definition">#</a></dt>
<dd><p>Call this to load a fitted model</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>pkl_file (str)</dt><dd><p>Location of saved .pkl file on disk</p>
</dd>
<dt>npz_file (str)</dt><dd><p>Location of saved .npz file on disk</p>
</dd>
<dt>outPrefix (str)</dt><dd><p>Output prefix for model to save to (e.g. plots)</p>
</dd>
<dt>max_samples (int)</dt><dd><p>Maximum samples if subsampling X
[default = 100000]</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to load npz file with GPU libraries
for lineage models</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>load_obj (model)</dt><dd><p>Loaded model</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="network-py">
<h2>network.py<a class="headerlink" href="#network-py" title="Permalink to this heading">#</a></h2>
<p>Functions used to construct the network, and update with new queries. Main
entry point is <code class="xref py py-func docutils literal notranslate"><span class="pre">constructNetwork()</span></code> for new reference
databases, and <code class="xref py py-func docutils literal notranslate"><span class="pre">findQueryLinksToNetwork()</span></code> for querying
databases.</p>
<span class="target" id="module-PopPUNK.network"></span><p>Network functions</p>
<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.addQueryToNetwork">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">addQueryToNetwork</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbFuncs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rList</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qList</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assignments</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">queryDB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kmers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'euclidean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">queryQuery</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_preserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#addQueryToNetwork"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.addQueryToNetwork" title="Permalink to this definition">#</a></dt>
<dd><p>Finds edges between queries and items in the reference database,
and modifies the network to include them.</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>dbFuncs (list)</dt><dd><p>List of backend functions from <a class="reference internal" href="#PopPUNK.utils.setupDBFuncs" title="PopPUNK.utils.setupDBFuncs"><code class="xref py py-func docutils literal notranslate"><span class="pre">setupDBFuncs()</span></code></a></p>
</dd>
<dt>rList (list)</dt><dd><p>List of reference names</p>
</dd>
<dt>qList (list)</dt><dd><p>List of query names</p>
</dd>
<dt>G (graph)</dt><dd><p>Network to add to (mutated)</p>
</dd>
<dt>assignments (numpy.array)</dt><dd><p>Cluster assignment of items in qlist</p>
</dd>
<dt>model (ClusterModel)</dt><dd><p>Model fitted to reference database</p>
</dd>
<dt>queryDB (str)</dt><dd><p>Query database location</p>
</dd>
<dt>distances (str)</dt><dd><p>Prefix of distance files for extending network</p>
</dd>
<dt>kmers (list)</dt><dd><p>List of k-mer sizes</p>
</dd>
<dt>distance_type (str)</dt><dd><p>Distance type to use as weights in network</p>
</dd>
<dt>queryQuery (bool)</dt><dd><p>Add in all query-query distances
(default = False)</p>
</dd>
<dt>strand_preserved (bool)</dt><dd><p>Whether to treat strand as known (i.e. ignore rc k-mers)
when adding random distances. Only used if queryQuery = True
[default = False]</p>
</dd>
<dt>weights (numpy.array)</dt><dd><p>If passed, the core,accessory distances for each assignment, which will
be annotated as an edge attribute</p>
</dd>
<dt>threads (int)</dt><dd><p>Number of threads to use if new db created</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use cugraph for analysis</p>
<p>(default = 1)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>distMat (numpy.array)</dt><dd><p>Query-query distances</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.add_self_loop">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">add_self_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_num</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">renumber</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#add_self_loop"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.add_self_loop" title="Permalink to this definition">#</a></dt>
<dd><p>Adds self-loop to cugraph graph to ensure all nodes are included in
the graph, even if singletons.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G_df (cudf)</dt><dd><p>cudf data frame containing edge list</p>
</dd>
<dt>seq_num (int)</dt><dd><p>The expected number of nodes in the graph</p>
</dd>
<dt>renumber (bool)</dt><dd><p>Whether to renumber the vertices when added to the graph</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>G_new (graph)</dt><dd><p>Dictionary of cluster assignments (keys are sequence names)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.checkNetworkVertexCount">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">checkNetworkVertexCount</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#checkNetworkVertexCount"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.checkNetworkVertexCount" title="Permalink to this definition">#</a></dt>
<dd><p>Checks the number of network vertices matches the number
of sequence names.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>seq_list (list)</dt><dd><p>The list of sequence names</p>
</dd>
<dt>G (graph)</dt><dd><p>The network of sequences</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use cugraph for graph analyses</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.cliquePrune">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">cliquePrune</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">component</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_indices</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">components_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#cliquePrune"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.cliquePrune" title="Permalink to this definition">#</a></dt>
<dd><p>Wrapper function around <a class="reference internal" href="#PopPUNK.network.getCliqueRefs" title="PopPUNK.network.getCliqueRefs"><code class="xref py py-func docutils literal notranslate"><span class="pre">getCliqueRefs()</span></code></a> so it can be
called by a multiprocessing pool</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.construct_dense_weighted_network">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">construct_dense_weighted_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distMat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#construct_dense_weighted_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.construct_dense_weighted_network" title="Permalink to this definition">#</a></dt>
<dd><p>Construct an undirected network using sequence lists, assignments of pairwise distances
to clusters, and the identifier of the cluster assigned to within-strain distances.
Nodes are samples and edges where samples are within the same cluster</p>
<p>Will print summary statistics about the network to <code class="docutils literal notranslate"><span class="pre">STDERR</span></code></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>rlist (list)</dt><dd><p>List of reference sequence labels</p>
</dd>
<dt>distMat (2 column ndarray)</dt><dd><p>Numpy array of pairwise distances</p>
</dd>
<dt>weights_type (str)</dt><dd><p>Type of weight to use for network</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use GPUs for network construction</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>G (graph)</dt><dd><p>The resulting network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.construct_network_from_assignments">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">construct_network_from_assignments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assignments</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">within_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distMat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">previous_network</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">old_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adding_qq_dists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">previous_pkl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">betweenness_sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summarise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#construct_network_from_assignments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.construct_network_from_assignments" title="Permalink to this definition">#</a></dt>
<dd><p>Construct an undirected network using sequence lists, assignments of pairwise distances
to clusters, and the identifier of the cluster assigned to within-strain distances.
Nodes are samples and edges where samples are within the same cluster</p>
<p>Will print summary statistics about the network to <code class="docutils literal notranslate"><span class="pre">STDERR</span></code></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>rlist (list)</dt><dd><p>List of reference sequence labels</p>
</dd>
<dt>qlist (list)</dt><dd><p>List of query sequence labels</p>
</dd>
<dt>assignments (numpy.array or int)</dt><dd><p>Labels of most likely cluster assignment</p>
</dd>
<dt>within_label (int)</dt><dd><p>The label for the cluster representing within-strain distances</p>
</dd>
<dt>int_offset (int)</dt><dd><p>Constant integer to add to each node index</p>
</dd>
<dt>weights (list)</dt><dd><p>List of weights for each edge in the network</p>
</dd>
<dt>distMat (2 column ndarray)</dt><dd><p>Numpy array of pairwise distances</p>
</dd>
<dt>weights_type (str)</dt><dd><p>Measure to calculate from the distMat to use as edge weights in network
- options are core, accessory or euclidean distance</p>
</dd>
<dt>previous_network (str)</dt><dd><p>Name of file containing a previous network to be integrated into this new
network</p>
</dd>
<dt>old_ids (list)</dt><dd><p>Ordered list of vertex names in previous network</p>
</dd>
<dt>adding_qq_dists (bool)</dt><dd><p>Boolean specifying whether query-query edges are being added
to an existing network, such that not all the sequence IDs will
be found in the old IDs, which should already be correctly ordered</p>
</dd>
<dt>previous_pkl (str)</dt><dd><p>Name of file containing the names of the sequences in the previous_network</p>
</dd>
<dt>betweenness_sample (int)</dt><dd><p>Number of sequences per component used to estimate betweenness using
a GPU. Smaller numbers are faster but less precise [default = 100]</p>
</dd>
<dt>summarise (bool)</dt><dd><p>Whether to calculate and print network summaries with <a class="reference internal" href="#PopPUNK.network.networkSummary" title="PopPUNK.network.networkSummary"><code class="xref py py-func docutils literal notranslate"><span class="pre">networkSummary()</span></code></a>
(default = True)</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use GPUs for network construction</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>G (graph)</dt><dd><p>The resulting network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.construct_network_from_df">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">construct_network_from_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">G_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distMat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">previous_network</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adding_qq_dists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">old_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">previous_pkl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">betweenness_sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summarise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#construct_network_from_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.construct_network_from_df" title="Permalink to this definition">#</a></dt>
<dd><p>Construct an undirected network using a data frame of edges. Nodes are samples and
edges where samples are within the same cluster</p>
<p>Will print summary statistics about the network to <code class="docutils literal notranslate"><span class="pre">STDERR</span></code></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>rlist (list)</dt><dd><p>List of reference sequence labels</p>
</dd>
<dt>qlist (list)</dt><dd><p>List of query sequence labels</p>
</dd>
<dt>G_df (cudf or pandas data frame)</dt><dd><p>Data frame in which the first two columns are the nodes linked by edges</p>
</dd>
<dt>weights (bool)</dt><dd><p>Whether weights in the G_df data frame should be included in the network</p>
</dd>
<dt>distMat (2 column ndarray)</dt><dd><p>Numpy array of pairwise distances</p>
</dd>
<dt>previous_network (str or graph object)</dt><dd><p>Name of file containing a previous network to be integrated into this new
network, or the already-loaded graph object</p>
</dd>
<dt>adding_qq_dists (bool)</dt><dd><p>Boolean specifying whether query-query edges are being added
to an existing network, such that not all the sequence IDs will
be found in the old IDs, which should already be correctly ordered</p>
</dd>
<dt>old_ids (list)</dt><dd><p>Ordered list of vertex names in previous network</p>
</dd>
<dt>previous_pkl (str)</dt><dd><p>Name of file containing the names of the sequences in the previous_network</p>
</dd>
<dt>betweenness_sample (int)</dt><dd><p>Number of sequences per component used to estimate betweenness using
a GPU. Smaller numbers are faster but less precise [default = 100]</p>
</dd>
<dt>summarise (bool)</dt><dd><p>Whether to calculate and print network summaries with <a class="reference internal" href="#PopPUNK.network.networkSummary" title="PopPUNK.network.networkSummary"><code class="xref py py-func docutils literal notranslate"><span class="pre">networkSummary()</span></code></a>
(default = True)</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use GPUs for network construction</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>G (graph)</dt><dd><p>The resulting network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.construct_network_from_edge_list">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">construct_network_from_edge_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distMat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">previous_network</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adding_qq_dists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">old_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">previous_pkl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">betweenness_sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summarise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#construct_network_from_edge_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.construct_network_from_edge_list" title="Permalink to this definition">#</a></dt>
<dd><p>Construct an undirected network using a data frame of edges. Nodes are samples and
edges where samples are within the same cluster</p>
<p>Will print summary statistics about the network to <code class="docutils literal notranslate"><span class="pre">STDERR</span></code></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>rlist (list)</dt><dd><p>List of reference sequence labels</p>
</dd>
<dt>qlist (list)</dt><dd><p>List of query sequence labels</p>
</dd>
<dt>G_df (cudf or pandas data frame)</dt><dd><p>Data frame in which the first two columns are the nodes linked by edges</p>
</dd>
<dt>weights (list)</dt><dd><p>List of edge weights</p>
</dd>
<dt>distMat (2 column ndarray)</dt><dd><p>Numpy array of pairwise distances</p>
</dd>
<dt>previous_network (str or graph object)</dt><dd><p>Name of file containing a previous network to be integrated into this new
network, or the already-loaded graph object</p>
</dd>
<dt>adding_qq_dists (bool)</dt><dd><p>Boolean specifying whether query-query edges are being added
to an existing network, such that not all the sequence IDs will
be found in the old IDs, which should already be correctly ordered</p>
</dd>
<dt>old_ids (list)</dt><dd><p>Ordered list of vertex names in previous network</p>
</dd>
<dt>previous_pkl (str)</dt><dd><p>Name of file containing the names of the sequences in the previous_network</p>
</dd>
<dt>betweenness_sample (int)</dt><dd><p>Number of sequences per component used to estimate betweenness using
a GPU. Smaller numbers are faster but less precise [default = 100]</p>
</dd>
<dt>summarise (bool)</dt><dd><p>Whether to calculate and print network summaries with <a class="reference internal" href="#PopPUNK.network.networkSummary" title="PopPUNK.network.networkSummary"><code class="xref py py-func docutils literal notranslate"><span class="pre">networkSummary()</span></code></a>
(default = True)</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use GPUs for network construction</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>G (graph)</dt><dd><p>The resulting network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.construct_network_from_sparse_matrix">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">construct_network_from_sparse_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sparse_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">previous_network</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">previous_pkl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">betweenness_sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summarise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#construct_network_from_sparse_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.construct_network_from_sparse_matrix" title="Permalink to this definition">#</a></dt>
<dd><p>Construct an undirected network using a sparse matrix. Nodes are samples and
edges where samples are within the same cluster</p>
<p>Will print summary statistics about the network to <code class="docutils literal notranslate"><span class="pre">STDERR</span></code></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>rlist (list)</dt><dd><p>List of reference sequence labels</p>
</dd>
<dt>qlist (list)</dt><dd><p>List of query sequence labels</p>
</dd>
<dt>sparse_input (numpy.array)</dt><dd><p>Sparse distance matrix from lineage fit</p>
</dd>
<dt>weights (list)</dt><dd><p>List of weights for each edge in the network</p>
</dd>
<dt>distMat (2 column ndarray)</dt><dd><p>Numpy array of pairwise distances</p>
</dd>
<dt>previous_network (str)</dt><dd><p>Name of file containing a previous network to be integrated into this new
network</p>
</dd>
<dt>previous_pkl (str)</dt><dd><p>Name of file containing the names of the sequences in the previous_network</p>
</dd>
<dt>betweenness_sample (int)</dt><dd><p>Number of sequences per component used to estimate betweenness using
a GPU. Smaller numbers are faster but less precise [default = 100]</p>
</dd>
<dt>summarise (bool)</dt><dd><p>Whether to calculate and print network summaries with <a class="reference internal" href="#PopPUNK.network.networkSummary" title="PopPUNK.network.networkSummary"><code class="xref py py-func docutils literal notranslate"><span class="pre">networkSummary()</span></code></a>
(default = True)</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use GPUs for network construction</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>G (graph)</dt><dd><p>The resulting network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.cugraph_to_graph_tool">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">cugraph_to_graph_tool</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rlist</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#cugraph_to_graph_tool"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.cugraph_to_graph_tool" title="Permalink to this definition">#</a></dt>
<dd><p>Save a network to disk</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G (cugraph network)</dt><dd><p>Cugraph network</p>
</dd>
<dt>rlist (list)</dt><dd><p>List of sequence names</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>G (graph-tool network)</dt><dd><p>Graph tool network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.extractReferences">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">extractReferences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dbOrder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outSuffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_isolate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">existingRefs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#extractReferences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.extractReferences" title="Permalink to this definition">#</a></dt>
<dd><p>Extract references for each cluster based on cliques</p>
<p>Writes chosen references to file by calling <a class="reference internal" href="#PopPUNK.network.writeReferences" title="PopPUNK.network.writeReferences"><code class="xref py py-func docutils literal notranslate"><span class="pre">writeReferences()</span></code></a></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G (graph)</dt><dd><p>A network used to define clusters</p>
</dd>
<dt>dbOrder (list)</dt><dd><p>The order of files in the sketches, so returned references are in the same order</p>
</dd>
<dt>outPrefix (str)</dt><dd><p>Prefix for output file</p>
</dd>
<dt>outSuffix (str)</dt><dd><p>Suffix for output file  (.refs will be appended)</p>
</dd>
<dt>type_isolate (str)</dt><dd><p>Isolate to be included in set of references</p>
</dd>
<dt>existingRefs (list)</dt><dd><p>References that should be used for each clique</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Use cugraph for graph analysis (default = False)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>refFileName (str)</dt><dd><p>The name of the file references were written to</p>
</dd>
<dt>references (list)</dt><dd><p>An updated list of the reference names</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.fetchNetwork">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">fetchNetwork</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refList</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">core_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accessory_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#fetchNetwork"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.fetchNetwork" title="Permalink to this definition">#</a></dt>
<dd><p>Load the network based on input options</p>
<p>Returns the network as a graph-tool format graph, and sets
the slope parameter of the passed model object.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>network_dir (str)</dt><dd><p>A network used to define clusters</p>
</dd>
<dt>model (ClusterFit)</dt><dd><p>A fitted model object</p>
</dd>
<dt>refList (list)</dt><dd><p>Names of references that should be in the network</p>
</dd>
<dt>ref_graph (bool)</dt><dd><p>Use ref only graph, if available
[default = False]</p>
</dd>
<dt>core_only (bool)</dt><dd><p>Return the network created using only core distances
[default = False]</p>
</dd>
<dt>accessory_only (bool)</dt><dd><p>Return the network created using only accessory distances
[default = False]</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Use cugraph library to load graph</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>genomeNetwork (graph)</dt><dd><p>The loaded network</p>
</dd>
<dt>cluster_file (str)</dt><dd><p>The CSV of cluster assignments corresponding to this network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.generate_minimum_spanning_tree">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">generate_minimum_spanning_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_cugraph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#generate_minimum_spanning_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.generate_minimum_spanning_tree" title="Permalink to this definition">#</a></dt>
<dd><p>Generate a minimum spanning tree from a network</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G (network)</dt><dd><p>Graph tool network</p>
</dd>
<dt>from_cugraph (bool)</dt><dd><p>If a pre-calculated MST from cugraph
[default = False]</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>mst_network (str)</dt><dd><p>Minimum spanning tree (as graph-tool graph)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.getCliqueRefs">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">getCliqueRefs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_indices</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#getCliqueRefs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.getCliqueRefs" title="Permalink to this definition">#</a></dt>
<dd><p>Recursively prune a network of its cliques. Returns one vertex from
a clique at each stage</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G (graph)</dt><dd><p>The graph to get clique representatives from</p>
</dd>
<dt>reference_indices (set)</dt><dd><p>The unique list of vertices being kept, to add to</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.get_cugraph_triangles">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">get_cugraph_triangles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#get_cugraph_triangles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.get_cugraph_triangles" title="Permalink to this definition">#</a></dt>
<dd><p>Counts the number of triangles in a cugraph
network. Can be removed when the cugraph issue
<a class="reference external" href="https://github.com/rapidsai/cugraph/issues/1043">https://github.com/rapidsai/cugraph/issues/1043</a> is fixed.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G (cugraph network)</dt><dd><p>Network to be analysed</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>triangle_count (int)</dt><dd><p>Count of triangles in graph</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.get_vertex_list">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">get_vertex_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#get_vertex_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.get_vertex_list" title="Permalink to this definition">#</a></dt>
<dd><p>Generate a list of node indices</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G (network)</dt><dd><p>Graph tool network</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether graph is a cugraph or not
[default = False]</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>vlist (list)</dt><dd><p>List of integers corresponding to nodes</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.load_network_file">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">load_network_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#load_network_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.load_network_file" title="Permalink to this definition">#</a></dt>
<dd><p>Load the network based on input options</p>
<p>Returns the network as a graph-tool format graph, and sets
the slope parameter of the passed model object.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>fn (str)</dt><dd><p>Network file name</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Use cugraph library to load graph</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>genomeNetwork (graph)</dt><dd><p>The loaded network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.networkSummary">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">networkSummary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calc_betweenness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">betweenness_sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#networkSummary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.networkSummary" title="Permalink to this definition">#</a></dt>
<dd><p>Provides summary values about the network</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G (graph)</dt><dd><p>The network of strains</p>
</dd>
<dt>calc_betweenness (bool)</dt><dd><p>Whether to calculate betweenness stats</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use cugraph for graph analysis</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>metrics (list)</dt><dd><p>List with # components, density, transitivity, mean betweenness
and weighted mean betweenness</p>
</dd>
<dt>scores (list)</dt><dd><p>List of scores</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.network_to_edges">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">network_to_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prev_G_fn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adding_qq_dists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">old_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">previous_pkl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#network_to_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.network_to_edges" title="Permalink to this definition">#</a></dt>
<dd><p>Load previous network, extract the edges to match the
vertex order specified in rlist, and also return weights if specified.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>prev_G_fn (str or graph object)</dt><dd><p>Path of file containing existing network, or already-loaded
graph object</p>
</dd>
<dt>adding_qq_dists (bool)</dt><dd><p>Boolean specifying whether query-query edges are being added
to an existing network, such that not all the sequence IDs will
be found in the old IDs, which should already be correctly ordered</p>
</dd>
<dt>rlist (list)</dt><dd><p>List of reference sequence labels in new network</p>
</dd>
<dt>old_ids (list)</dt><dd><p>List of IDs of vertices in existing network</p>
</dd>
<dt>previous_pkl (str)</dt><dd><p>Path of pkl file containing names of sequences in
previous network</p>
</dd>
<dt>weights (bool)</dt><dd><p>Whether to return edge weights
(default = False)</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use cugraph for graph analyses</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>source_ids (list)</dt><dd><p>Source nodes for each edge</p>
</dd>
<dt>target_ids (list)</dt><dd><p>Target nodes for each edge</p>
</dd>
<dt>edge_weights (list)</dt><dd><p>Weights for each new edge</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.printClusters">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">printClusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">oldClusterFile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">externalClusterCSV</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">printRef</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">printCSV</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'combined'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_unwords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#printClusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.printClusters" title="Permalink to this definition">#</a></dt>
<dd><p>Get cluster assignments</p>
<p>Also writes assignments to a CSV file</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G (graph)</dt><dd><p>Network used to define clusters</p>
</dd>
<dt>rlist (list)</dt><dd><p>Names of samples</p>
</dd>
<dt>outPrefix (str)</dt><dd><p>Prefix for output CSV
Default = None</p>
</dd>
<dt>oldClusterFile (str)</dt><dd><p>CSV with previous cluster assignments.
Pass to ensure consistency in cluster assignment name.
Default = None</p>
</dd>
<dt>externalClusterCSV (str)</dt><dd><p>CSV with cluster assignments from any source. Will print a file
relating these to new cluster assignments
Default = None</p>
</dd>
<dt>printRef (bool)</dt><dd><p>If false, print only query sequences in the output
Default = True</p>
</dd>
<dt>printCSV (bool)</dt><dd><p>Print results to file
Default = True</p>
</dd>
<dt>clustering_type (str)</dt><dd><p>Type of clustering network, used for comparison with old clusters
Default = ‘combined’</p>
</dd>
<dt>write_unwords (bool)</dt><dd><p>Write clusters with a pronouncable name rather than numerical index
Default = True</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use cugraph for network analysis</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>clustering (dict)</dt><dd><p>Dictionary of cluster assignments (keys are sequence names)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.printExternalClusters">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">printExternalClusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">newClusters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extClusterFile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">oldNames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">printRef</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#printExternalClusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.printExternalClusters" title="Permalink to this definition">#</a></dt>
<dd><p>Prints cluster assignments with respect to previously defined
clusters or labels.</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>newClusters (set iterable)</dt><dd><p>The components from the graph G, defining the PopPUNK clusters</p>
</dd>
<dt>extClusterFile (str)</dt><dd><p>A CSV file containing definitions of the external clusters for
each sample (does not need to contain all samples)</p>
</dd>
<dt>outPrefix (str)</dt><dd><p>Prefix for output CSV (_external_clusters.csv)</p>
</dd>
<dt>oldNames (list)</dt><dd><p>A list of the reference sequences</p>
</dd>
<dt>printRef (bool)</dt><dd><p>If false, print only query sequences in the output</p>
<p>Default = True</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.print_network_summary">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">print_network_summary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">betweenness_sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#print_network_summary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.print_network_summary" title="Permalink to this definition">#</a></dt>
<dd><p>Wrapper function for printing network information</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G (graph)</dt><dd><p>List of reference sequence labels</p>
</dd>
<dt>betweenness_sample (int)</dt><dd><p>Number of sequences per component used to estimate betweenness using
a GPU. Smaller numbers are faster but less precise [default = 100]</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use GPUs for network construction</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.process_previous_network">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">process_previous_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">previous_network</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adding_qq_dists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">old_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">previous_pkl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertex_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#process_previous_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.process_previous_network" title="Permalink to this definition">#</a></dt>
<dd><p>Extract edge types from an existing network</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>previous_network (str or graph object)</dt><dd><p>Name of file containing a previous network to be integrated into this new
network, or already-loaded graph object</p>
</dd>
<dt>adding_qq_dists (bool)</dt><dd><p>Boolean specifying whether query-query edges are being added
to an existing network, such that not all the sequence IDs will
be found in the old IDs, which should already be correctly ordered</p>
</dd>
<dt>old_ids (list)</dt><dd><p>Ordered list of vertex names in previous network</p>
</dd>
<dt>previous_pkl (str)</dt><dd><p>Name of file containing the names of the sequences in the previous_network
ordered based on the original network construction</p>
</dd>
<dt>vertex_labels (list)</dt><dd><p>Ordered list of sequence labels</p>
</dd>
<dt>weights (bool)</dt><dd><p>Whether weights should be extracted from the previous network</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use GPUs for network construction</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>extra_sources (list)</dt><dd><p>List of source node identifiers</p>
</dd>
<dt>extra_targets (list)</dt><dd><p>List of destination node identifiers</p>
</dd>
<dt>extra_weights (list or None)</dt><dd><p>List of edge weights</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.process_weights">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">process_weights</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distMat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights_type</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#process_weights"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.process_weights" title="Permalink to this definition">#</a></dt>
<dd><p>Calculate edge weights from the distance matrix</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>distMat (2 column ndarray)</dt><dd><p>Numpy array of pairwise distances</p>
</dd>
<dt>weights_type (str)</dt><dd><p>Measure to calculate from the distMat to use as edge weights in network
- options are core, accessory or euclidean distance</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>processed_weights (list)</dt><dd><p>Edge weights</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.save_network">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">save_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_graphml</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#save_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.save_network" title="Permalink to this definition">#</a></dt>
<dd><p>Save a network to disk</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G (network)</dt><dd><p>Graph tool network</p>
</dd>
<dt>prefix (str)</dt><dd><p>Prefix for output file</p>
</dd>
<dt>use_graphml (bool)</dt><dd><p>Whether to output a graph-tool file
in graphml format</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether graph is a cugraph or not
[default = False]</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.sparse_mat_to_network">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">sparse_mat_to_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sparse_mat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#sparse_mat_to_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.sparse_mat_to_network" title="Permalink to this definition">#</a></dt>
<dd><p>Generate a network from a lineage rank fit</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>sparse_mat (scipy or cupyx sparse matrix)</dt><dd><p>Sparse matrix of kNN from lineage fit</p>
</dd>
<dt>rlist (list)</dt><dd><p>List of sequence names</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether GPU libraries should be used</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>G (network)</dt><dd><p>Graph tool or cugraph network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.translate_network_indices">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">translate_network_indices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G_ref_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_indices</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#translate_network_indices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.translate_network_indices" title="Permalink to this definition">#</a></dt>
<dd><p>Function for ensuring an updated reference network retains
numbering consistent with sample names</p>
<blockquote>
<div><dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G_ref_df (cudf data frame)</dt><dd><p>List of edges in reference network</p>
</dd>
<dt>reference_indices (list)</dt><dd><p>The ordered list of reference indices in the original network</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>G_ref (cugraph network)</dt><dd><p>Network of reference sequences</p>
</dd>
</dl>
</dd>
</dl>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.vertex_betweenness">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">vertex_betweenness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#vertex_betweenness"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.vertex_betweenness" title="Permalink to this definition">#</a></dt>
<dd><p>Returns betweenness for nodes in the graph</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.network.writeReferences">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.network.</span></span><span class="sig-name descname"><span class="pre">writeReferences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">refList</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outSuffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/network.html#writeReferences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.network.writeReferences" title="Permalink to this definition">#</a></dt>
<dd><p>Writes chosen references to file</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>refList (list)</dt><dd><p>Reference names to write</p>
</dd>
<dt>outPrefix (str)</dt><dd><p>Prefix for output file</p>
</dd>
<dt>outSuffix (str)</dt><dd><p>Suffix for output file (.refs will be appended)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>refFileName (str)</dt><dd><p>The name of the file references were written to</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="refine-py">
<h2>refine.py<a class="headerlink" href="#refine-py" title="Permalink to this heading">#</a></h2>
<p>Functions used to refine an existing model. Access using
<a class="reference internal" href="#PopPUNK.models.RefineFit" title="PopPUNK.models.RefineFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">RefineFit</span></code></a>.</p>
<span class="target" id="module-PopPUNK.refine"></span><p>Refine mixture model using network properties</p>
<dl class="py class">
<dt class="sig sig-object py" id="PopPUNK.refine.NumpyShared">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PopPUNK.refine.</span></span><span class="sig-name descname"><span class="pre">NumpyShared</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shape</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#PopPUNK.refine.NumpyShared" title="Permalink to this definition">#</a></dt>
<dd><dl class="py attribute">
<dt class="sig sig-object py" id="PopPUNK.refine.NumpyShared.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#PopPUNK.refine.NumpyShared.dtype" title="Permalink to this definition">#</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="PopPUNK.refine.NumpyShared.name">
<span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#PopPUNK.refine.NumpyShared.name" title="Permalink to this definition">#</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="PopPUNK.refine.NumpyShared.shape">
<span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#PopPUNK.refine.NumpyShared.shape" title="Permalink to this definition">#</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.refine.check_search_range">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.refine.</span></span><span class="sig-name descname"><span class="pre">check_search_range</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upper_s</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/refine.html#check_search_range"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.refine.check_search_range" title="Permalink to this definition">#</a></dt>
<dd><p>Checks a search range is within a valid range</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>scale (np.array)</dt><dd><p>Rescaling factor to [0, 1] for each axis</p>
</dd>
<dt>mean0 (np.array)</dt><dd><p>(x, y) of starting point defining line</p>
</dd>
<dt>mean1 (np.array)</dt><dd><p>(x, y) of end point defining line</p>
</dd>
<dt>lower_s (float)</dt><dd><p>distance along line to start search</p>
</dd>
<dt>upper_s (float)</dt><dd><p>distance along line to end search</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>min_x, max_x</dt><dd><p>minimum and maximum x-intercepts of the search range</p>
</dd>
<dt>min_y, max_y</dt><dd><p>minimum and maximum x-intercepts of the search range</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.refine.expand_cugraph_network">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.refine.</span></span><span class="sig-name descname"><span class="pre">expand_cugraph_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">G_extra_df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/refine.html#expand_cugraph_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.refine.expand_cugraph_network" title="Permalink to this definition">#</a></dt>
<dd><p>Reconstruct a cugraph network with additional edges.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G (cugraph network)</dt><dd><p>Original cugraph network</p>
</dd>
<dt>extra_edges (cudf dataframe)</dt><dd><p>Data frame of edges to add</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>G (cugraph network)</dt><dd><p>Expanded cugraph network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.refine.growNetwork">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.refine.</span></span><span class="sig-name descname"><span class="pre">growNetwork</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i_vec</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j_vec</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx_vec</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s_range</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thread_idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">betweenness_sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/refine.html#growNetwork"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.refine.growNetwork" title="Permalink to this definition">#</a></dt>
<dd><p>Construct a network, then add edges to it iteratively.
Input is from <code class="docutils literal notranslate"><span class="pre">pp_sketchlib.iterateBoundary1D</span></code> or``pp_sketchlib.iterateBoundary2D``</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>sample_names (list)</dt><dd><p>Sample names corresponding to distMat (accessed by iterator)</p>
</dd>
<dt>i_vec (list)</dt><dd><p>Ordered ref vertex index to add</p>
</dd>
<dt>j_vec (list)</dt><dd><p>Ordered query (==ref) vertex index to add</p>
</dd>
<dt>idx_vec (list)</dt><dd><p>For each i, j tuple, the index of the intercept at which these enter
the network. These are sorted and increasing</p>
</dd>
<dt>s_range (list)</dt><dd><p>Offsets which correspond to idx_vec entries</p>
</dd>
<dt>score_idx (int)</dt><dd><p>Index of score from <a class="reference internal" href="#PopPUNK.network.networkSummary" title="PopPUNK.network.networkSummary"><code class="xref py py-func docutils literal notranslate"><span class="pre">networkSummary()</span></code></a> to use
[default = 0]</p>
</dd>
<dt>thread_idx (int)</dt><dd><p>Optional thread idx (if multithreaded) to offset progress bar by</p>
</dd>
<dt>betweenness_sample (int)</dt><dd><p>Number of sequences per component used to estimate betweenness using
a GPU. Smaller numbers are faster but less precise [default = 100]</p>
</dd>
<dt>write_clusters (str)</dt><dd><p>Set to a prefix to write the clusters from each position to files
[default = None]</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use cugraph for graph analyses</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>scores (list)</dt><dd><p>-1 * network score for each of x_range.
Where network score is from <a class="reference internal" href="#PopPUNK.network.networkSummary" title="PopPUNK.network.networkSummary"><code class="xref py py-func docutils literal notranslate"><span class="pre">networkSummary()</span></code></a></p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.refine.likelihoodBoundary">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.refine.</span></span><span class="sig-name descname"><span class="pre">likelihoodBoundary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">within</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">between</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/refine.html#likelihoodBoundary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.refine.likelihoodBoundary" title="Permalink to this definition">#</a></dt>
<dd><p>Wrapper function around <a class="reference internal" href="#PopPUNK.bgmm.fit2dMultiGaussian" title="PopPUNK.bgmm.fit2dMultiGaussian"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit2dMultiGaussian()</span></code></a> so that it can
go into a root-finding function for probabilities between components</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>s (float)</dt><dd><p>Distance along line from mean0</p>
</dd>
<dt>model (BGMMFit)</dt><dd><p>Fitted mixture model</p>
</dd>
<dt>start (numpy.array)</dt><dd><p>The co-ordinates of the centre of the within-strain distribution</p>
</dd>
<dt>end (numpy.array)</dt><dd><p>The co-ordinates of the centre of the between-strain distribution</p>
</dd>
<dt>within (int)</dt><dd><p>Label of the within-strain distribution</p>
</dd>
<dt>between (int)</dt><dd><p>Label of the between-strain distribution</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>responsibility (float)</dt><dd><p>The difference between responsibilities of assignment to the within component
and the between assignment</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.refine.multi_refine">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.refine.</span></span><span class="sig-name descname"><span class="pre">multi_refine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distMat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s_max</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_boundary_points</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_processes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/refine.html#multi_refine"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.refine.multi_refine" title="Permalink to this definition">#</a></dt>
<dd><p>Move the refinement boundary between the optimum and where it meets an
axis. Discrete steps, output the clusers at each step</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>distMat (numpy.array)</dt><dd><p>n x 2 array of core and accessory distances for n samples</p>
</dd>
<dt>sample_names (list)</dt><dd><p>List of query sequence labels</p>
</dd>
<dt>mean0 (numpy.array)</dt><dd><p>Start point to define search line</p>
</dd>
<dt>mean1 (numpy.array)</dt><dd><p>End point to define search line</p>
</dd>
<dt>scale (numpy.array)</dt><dd><p>Scaling factor of distMat</p>
</dd>
<dt>s_max (float)</dt><dd><p>The optimal s position from refinement (<a class="reference internal" href="#PopPUNK.refine.refineFit" title="PopPUNK.refine.refineFit"><code class="xref py py-func docutils literal notranslate"><span class="pre">refineFit()</span></code></a>)</p>
</dd>
<dt>n_boundary_points (int)</dt><dd><p>Number of positions to try drawing the boundary at</p>
</dd>
<dt>num_processes (int)</dt><dd><p>Number of threads to use in the global optimisation step.
(default = 1)</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use cugraph for graph analyses</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.refine.newNetwork">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.refine.</span></span><span class="sig-name descname"><span class="pre">newNetwork</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distMat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gradient</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slope</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">betweenness_sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/refine.html#newNetwork"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.refine.newNetwork" title="Permalink to this definition">#</a></dt>
<dd><p>Wrapper function for <a class="reference internal" href="#PopPUNK.network.construct_network_from_edge_list" title="PopPUNK.network.construct_network_from_edge_list"><code class="xref py py-func docutils literal notranslate"><span class="pre">construct_network_from_edge_list()</span></code></a> which is called
by optimisation functions moving a triangular decision boundary.</p>
<p>Given the boundary parameterisation, constructs the network and returns
its score, to be minimised.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>s (float)</dt><dd><p>Distance along line between start_point and mean1 from start_point</p>
</dd>
<dt>sample_names (list)</dt><dd><p>Sample names corresponding to distMat (accessed by iterator)</p>
</dd>
<dt>distMat (numpy.array or NumpyShared)</dt><dd><p>Core and accessory distances or NumpyShared describing these in sharedmem</p>
</dd>
<dt>mean0 (numpy.array)</dt><dd><p>Start point</p>
</dd>
<dt>mean1 (numpy.array)</dt><dd><p>End point</p>
</dd>
<dt>gradient (float)</dt><dd><p>Gradient of line to move along</p>
</dd>
<dt>slope (int)</dt><dd><p>Set to 0 for a vertical line, 1 for a horizontal line, or
2 to use a slope
[default = 2]</p>
</dd>
<dt>score_idx (int)</dt><dd><p>Index of score from <a class="reference internal" href="#PopPUNK.network.networkSummary" title="PopPUNK.network.networkSummary"><code class="xref py py-func docutils literal notranslate"><span class="pre">networkSummary()</span></code></a> to use
[default = 0]</p>
</dd>
<dt>cpus (int)</dt><dd><p>Number of CPUs to use for calculating assignment</p>
</dd>
<dt>betweenness_sample (int)</dt><dd><p>Number of sequences per component used to estimate betweenness using
a GPU. Smaller numbers are faster but less precise [default = 100]</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use cugraph for graph analysis</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>score (float)</dt><dd><p>-1 * network score. Where network score is from <a class="reference internal" href="#PopPUNK.network.networkSummary" title="PopPUNK.network.networkSummary"><code class="xref py py-func docutils literal notranslate"><span class="pre">networkSummary()</span></code></a></p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.refine.newNetwork2D">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.refine.</span></span><span class="sig-name descname"><span class="pre">newNetwork2D</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y_idx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distMat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_range</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_range</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">betweenness_sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/refine.html#newNetwork2D"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.refine.newNetwork2D" title="Permalink to this definition">#</a></dt>
<dd><p>Wrapper function for thresholdIterate2D and <a class="reference internal" href="#PopPUNK.refine.growNetwork" title="PopPUNK.refine.growNetwork"><code class="xref py py-func docutils literal notranslate"><span class="pre">growNetwork()</span></code></a>.</p>
<p>For a given y_max, constructs networks across x_range and returns a list
of scores</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>y_idx (float)</dt><dd><p>Maximum y-intercept of boundary, as index into y_range</p>
</dd>
<dt>sample_names (list)</dt><dd><p>Sample names corresponding to distMat (accessed by iterator)</p>
</dd>
<dt>distMat (numpy.array or NumpyShared)</dt><dd><p>Core and accessory distances or NumpyShared describing these in sharedmem</p>
</dd>
<dt>x_range (list)</dt><dd><p>Sorted list of x-intercepts to search</p>
</dd>
<dt>y_range (list)</dt><dd><p>Sorted list of y-intercepts to search</p>
</dd>
<dt>score_idx (int)</dt><dd><p>Index of score from <a class="reference internal" href="#PopPUNK.network.networkSummary" title="PopPUNK.network.networkSummary"><code class="xref py py-func docutils literal notranslate"><span class="pre">networkSummary()</span></code></a> to use
[default = 0]</p>
</dd>
<dt>betweenness_sample (int)</dt><dd><p>Number of sequences per component used to estimate betweenness using
a GPU. Smaller numbers are faster but less precise [default = 100]</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use cugraph for graph analysis</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>scores (list)</dt><dd><p>-1 * network score for each of x_range.
Where network score is from <a class="reference internal" href="#PopPUNK.network.networkSummary" title="PopPUNK.network.networkSummary"><code class="xref py py-func docutils literal notranslate"><span class="pre">networkSummary()</span></code></a></p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.refine.readManualStart">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.refine.</span></span><span class="sig-name descname"><span class="pre">readManualStart</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">startFile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/refine.html#readManualStart"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.refine.readManualStart" title="Permalink to this definition">#</a></dt>
<dd><p>Reads a file to define a manual start point, rather than using <code class="docutils literal notranslate"><span class="pre">--fit-model</span></code></p>
<p>Throws and exits if incorrectly formatted.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>startFile (str)</dt><dd><p>Name of file with values to read</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>mean0 (numpy.array)</dt><dd><p>Centre of within-strain distribution</p>
</dd>
<dt>mean1 (numpy.array)</dt><dd><p>Centre of between-strain distribution</p>
</dd>
<dt>scaled (bool)</dt><dd><p>True if means are scaled between [0,1]</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.refine.refineFit">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.refine.</span></span><span class="sig-name descname"><span class="pre">refineFit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distMat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_move</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_move</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slope</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unconstrained</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_local</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_processes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">betweenness_sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/refine.html#refineFit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.refine.refineFit" title="Permalink to this definition">#</a></dt>
<dd><p>Try to refine a fit by maximising a network score based on transitivity and density.</p>
<p>Iteratively move the decision boundary to do this, using starting point from existing model.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>distMat (numpy.array)</dt><dd><p>n x 2 array of core and accessory distances for n samples</p>
</dd>
<dt>sample_names (list)</dt><dd><p>List of query sequence labels</p>
</dd>
<dt>mean0 (numpy.array)</dt><dd><p>Start point to define search line</p>
</dd>
<dt>mean1 (numpy.array)</dt><dd><p>End point to define search line</p>
</dd>
<dt>scale (numpy.array)</dt><dd><p>Scaling factor of distMat</p>
</dd>
<dt>max_move (float)</dt><dd><p>Maximum distance to move away from start point</p>
</dd>
<dt>min_move (float)</dt><dd><p>Minimum distance to move away from start point</p>
</dd>
<dt>slope (int)</dt><dd><p>Set to 0 for a vertical line, 1 for a horizontal line, or
2 to use a slope</p>
</dd>
<dt>score_idx (int)</dt><dd><p>Index of score from <a class="reference internal" href="#PopPUNK.network.networkSummary" title="PopPUNK.network.networkSummary"><code class="xref py py-func docutils literal notranslate"><span class="pre">networkSummary()</span></code></a> to use
[default = 0]</p>
</dd>
<dt>unconstrained (bool)</dt><dd><p>If True, search in 2D and change the slope of the boundary</p>
</dd>
<dt>no_local (bool)</dt><dd><p>Turn off the local optimisation step.
Quicker, but may be less well refined.</p>
</dd>
<dt>num_processes (int)</dt><dd><p>Number of threads to use in the global optimisation step.
(default = 1)</p>
</dd>
<dt>betweenness_sample (int)</dt><dd><p>Number of sequences per component used to estimate betweenness using
a GPU. Smaller numbers are faster but less precise [default = 100]</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use cugraph for graph analyses</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>optimal_x (float)</dt><dd><p>x-coordinate of refined fit</p>
</dd>
<dt>optimal_y (float)</dt><dd><p>y-coordinate of refined fit</p>
</dd>
<dt>optimised_s (float)</dt><dd><p>Position along search range of refined fit</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-PopPUNK.plot">
<span id="plot-py"></span><h2>plot.py<a class="headerlink" href="#module-PopPUNK.plot" title="Permalink to this heading">#</a></h2>
<p>Plots of GMM results, k-mer fits, and microreact output</p>
<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.createMicroreact">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">createMicroreact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">microreact_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">api_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#createMicroreact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.createMicroreact" title="Permalink to this definition">#</a></dt>
<dd><p>Creates a .microreact file, and instance via the API</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>prefix (str)</dt><dd><p>Prefix for output file</p>
</dd>
<dt>microreact_files (str)</dt><dd><p>List of Microreact files [clusters, dot, tree, mst_tree]</p>
</dd>
<dt>api_key (str)</dt><dd><p>API key for your account</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.distHistogram">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">distHistogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dists</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#distHistogram"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.distHistogram" title="Permalink to this definition">#</a></dt>
<dd><p>Plot a histogram of distances (1D)</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>dists (np.array)</dt><dd><p>Distance vector</p>
</dd>
<dt>rank (int)</dt><dd><p>Rank (used for name and title)</p>
</dd>
<dt>outPrefix (int)</dt><dd><p>Full path prefix for plot file</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.drawMST">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">drawMST</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mst</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isolate_clustering</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#drawMST"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.drawMST" title="Permalink to this definition">#</a></dt>
<dd><p>Plot a layout of the minimum spanning tree</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>mst (graph_tool.Graph)</dt><dd><p>A minimum spanning tree</p>
</dd>
<dt>outPrefix (str)</dt><dd><p>Output prefix for save files</p>
</dd>
<dt>isolate_clustering (dict)</dt><dd><p>Dictionary of ID: cluster, used for colouring vertices</p>
</dd>
<dt>clustering_name (str)</dt><dd><p>Name of clustering scheme to be used for colouring</p>
</dd>
<dt>overwrite (bool)</dt><dd><p>Overwrite existing output files</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.get_grid">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">get_grid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">minimum</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maximum</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#get_grid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.get_grid" title="Permalink to this definition">#</a></dt>
<dd><p>Get a square grid of points to evaluate a function across</p>
<p>Used for <a class="reference internal" href="#PopPUNK.plot.plot_scatter" title="PopPUNK.plot.plot_scatter"><code class="xref py py-func docutils literal notranslate"><span class="pre">plot_scatter()</span></code></a> and <a class="reference internal" href="#PopPUNK.plot.plot_contours" title="PopPUNK.plot.plot_contours"><code class="xref py py-func docutils literal notranslate"><span class="pre">plot_contours()</span></code></a></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>minimum (float)</dt><dd><p>Minimum value for grid</p>
</dd>
<dt>maximum (float)</dt><dd><p>Maximum value for grid</p>
</dd>
<dt>resolution (int)</dt><dd><p>Number of points along each axis</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>xx (numpy.array)</dt><dd><p>x values across n x n grid</p>
</dd>
<dt>yy (numpy.array)</dt><dd><p>y values across n x n grid</p>
</dd>
<dt>xy (numpy.array)</dt><dd><p>n x 2 pairs of x, y values grid is over</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.outputsForCytoscape">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">outputsForCytoscape</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">G_mst</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isolate_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epiCsv</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">queryList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">writeCsv</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#outputsForCytoscape"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.outputsForCytoscape" title="Permalink to this definition">#</a></dt>
<dd><p>Write outputs for cytoscape. A graphml of the network, and CSV with metadata</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>G (graph)</dt><dd><p>The network to write</p>
</dd>
<dt>G_mst (graph)</dt><dd><p>The minimum spanning tree of G</p>
</dd>
<dt>isolate_names (list)</dt><dd><p>Ordered list of sequence names</p>
</dd>
<dt>clustering (dict)</dt><dd><p>Dictionary of cluster assignments (keys are nodeNames).</p>
</dd>
<dt>outPrefix (str)</dt><dd><p>Prefix for files to be written</p>
</dd>
<dt>epiCsv (str)</dt><dd><p>Optional CSV of epi data to paste in the output in addition to
the clusters.</p>
</dd>
<dt>queryList (list)</dt><dd><p>Optional list of isolates that have been added as a query.
(default = None)</p>
</dd>
<dt>suffix (string)</dt><dd><p>String to append to network file name.
(default = None)</p>
</dd>
<dt>writeCsv (bool)</dt><dd><p>Whether to print CSV file to accompany network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.outputsForGrapetree">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">outputsForGrapetree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">combined_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nj_tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mst_tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epiCsv</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">queryList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#outputsForGrapetree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.outputsForGrapetree" title="Permalink to this definition">#</a></dt>
<dd><p>Generate files for Grapetree</p>
<p>Write a neighbour joining tree (.nwk) from core distances
and cluster assignment (.csv)</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>combined_list (list)</dt><dd><p>Name of sequences being analysed. The part of the name before the
first ‘.’ will be shown in the output</p>
</dd>
<dt>clustering (dict or dict of dicts)</dt><dd><p>List of cluster assignments from <a class="reference internal" href="#PopPUNK.network.printClusters" title="PopPUNK.network.printClusters"><code class="xref py py-func docutils literal notranslate"><span class="pre">printClusters()</span></code></a>.
Further clusterings (e.g. 1D core only) can be included by passing these
as a dict.</p>
</dd>
<dt>nj_tree (str or None)</dt><dd><p>String representation of a Newick-formatted NJ tree</p>
</dd>
<dt>mst_tree (str or None)</dt><dd><p>String representation of a Newick-formatted minimum-spanning tree</p>
</dd>
<dt>outPrefix (str)</dt><dd><p>Prefix for all generated output files, which will be placed in <cite>outPrefix</cite>
subdirectory.</p>
</dd>
<dt>epiCsv (str)</dt><dd><p>A CSV containing other information, to include with the CSV of clusters</p>
</dd>
<dt>queryList (list)</dt><dd><p>Optional list of isolates that have been added as a query for colouring
in the CSV. (default = None)</p>
</dd>
<dt>overwrite (bool)</dt><dd><p>Overwrite existing output if present (default = False).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.outputsForMicroreact">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">outputsForMicroreact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">combined_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nj_tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mst_tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accMat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perplexity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxIter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epiCsv</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">queryList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#outputsForMicroreact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.outputsForMicroreact" title="Permalink to this definition">#</a></dt>
<dd><p>Generate files for microreact</p>
<p>Output a neighbour joining tree (.nwk) from core distances, a plot of t-SNE clustering
of accessory distances (.dot) and cluster assignment (.csv)</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>combined_list (list)</dt><dd><p>Name of sequences being analysed. The part of the name before the first ‘.’ will
be shown in the output</p>
</dd>
<dt>clustering (dict or dict of dicts)</dt><dd><p>List of cluster assignments from <a class="reference internal" href="#PopPUNK.network.printClusters" title="PopPUNK.network.printClusters"><code class="xref py py-func docutils literal notranslate"><span class="pre">printClusters()</span></code></a>.
Further clusterings (e.g. 1D core only) can be included by passing these as a dict.</p>
</dd>
<dt>nj_tree (str or None)</dt><dd><p>String representation of a Newick-formatted NJ tree</p>
</dd>
<dt>mst_tree (str or None)</dt><dd><p>String representation of a Newick-formatted minimum-spanning tree</p>
</dd>
<dt>accMat (numpy.array)</dt><dd><p>n x n array of accessory distances for n samples.</p>
</dd>
<dt>perplexity (int)</dt><dd><p>Perplexity parameter passed to mandrake</p>
</dd>
<dt>maxIter (int)</dt><dd><p>Maximum iterations for mandrake</p>
</dd>
<dt>outPrefix (str)</dt><dd><p>Prefix for all generated output files, which will be placed in <cite>outPrefix</cite> subdirectory</p>
</dd>
<dt>epiCsv (str)</dt><dd><p>A CSV containing other information, to include with the CSV of clusters</p>
</dd>
<dt>queryList (list)</dt><dd><p>Optional list of isolates that have been added as a query for colouring in the CSV.
(default = None)</p>
</dd>
<dt>overwrite (bool)</dt><dd><p>Overwrite existing output if present (default = False)</p>
</dd>
<dt>n_threads (int)</dt><dd><p>Number of CPU threads to use
(default = 1)</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Whether to use a GPU for t-SNE generation</p>
</dd>
<dt>device_id (int)</dt><dd><p>Device ID of GPU to be used
(default = 0)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>outfiles (list)</dt><dd><p>List of output files create</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.outputsForPhandango">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">outputsForPhandango</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">combined_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nj_tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mst_tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epiCsv</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">queryList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#outputsForPhandango"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.outputsForPhandango" title="Permalink to this definition">#</a></dt>
<dd><p>Generate files for Phandango</p>
<p>Write a neighbour joining tree (.tree) from core distances
and cluster assignment (.csv)</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>combined_list (list)</dt><dd><p>Name of sequences being analysed. The part of the name before the first ‘.’ will
be shown in the output</p>
</dd>
<dt>clustering (dict or dict of dicts)</dt><dd><p>List of cluster assignments from <a class="reference internal" href="#PopPUNK.network.printClusters" title="PopPUNK.network.printClusters"><code class="xref py py-func docutils literal notranslate"><span class="pre">printClusters()</span></code></a>.
Further clusterings (e.g. 1D core only) can be included by passing these as a dict.</p>
</dd>
<dt>nj_tree (str or None)</dt><dd><p>String representation of a Newick-formatted NJ tree</p>
</dd>
<dt>mst_tree (str or None)</dt><dd><p>String representation of a Newick-formatted minimum-spanning tree</p>
</dd>
<dt>outPrefix (str)</dt><dd><p>Prefix for all generated output files, which will be placed in <cite>outPrefix</cite> subdirectory</p>
</dd>
<dt>epiCsv (str)</dt><dd><p>A CSV containing other information, to include with the CSV of clusters</p>
</dd>
<dt>queryList (list)</dt><dd><p>Optional list of isolates that have been added as a query for colouring in the CSV.
(default = None)</p>
</dd>
<dt>overwrite (bool)</dt><dd><p>Overwrite existing output if present (default = False)</p>
</dd>
<dt>threads (int)</dt><dd><p>Number of threads to use with rapidnj</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.plot_contours">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">plot_contours</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assignments</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_prefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#plot_contours"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.plot_contours" title="Permalink to this definition">#</a></dt>
<dd><p>Draw contours of mixture model assignments</p>
<p>Will draw the decision boundary for between/within in red</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>model (BGMMFit)</dt><dd><p>Model we are plotting from</p>
</dd>
<dt>assignments (numpy.array)</dt><dd><p>n-vectors of cluster assignments for model</p>
</dd>
<dt>title (str)</dt><dd><p>The title to display above the plot</p>
</dd>
<dt>out_prefix (str)</dt><dd><p>Prefix for output plot file (.pdf will be appended)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.plot_dbscan_results">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">plot_dbscan_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_clusters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_prefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#plot_dbscan_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.plot_dbscan_results" title="Permalink to this definition">#</a></dt>
<dd><p>Draw a scatter plot (png) to show the DBSCAN model fit</p>
<p>A scatter plot of core and accessory distances, coloured by component
membership. Black is noise</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>n x 2 array of core and accessory distances for n samples.</p>
</dd>
<dt>Y (numpy.array)</dt><dd><p>n x 1 array of cluster assignments for n samples.</p>
</dd>
<dt>n_clusters (int)</dt><dd><p>Number of clusters used (excluding noise)</p>
</dd>
<dt>out_prefix (str)</dt><dd><p>Prefix for output file (.png will be appended)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.plot_fit">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">plot_fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">klist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raw_matching</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raw_fit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corrected_matching</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corrected_fit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#plot_fit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.plot_fit" title="Permalink to this definition">#</a></dt>
<dd><p>Draw a scatter plot (pdf) of k-mer sizes vs match probability, and the
fit used to assign core and accessory distance</p>
<p>K-mer sizes on x-axis, log(pr(match)) on y - expect a straight line fit
with intercept representing accessory distance and slope core distance</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>klist (list)</dt><dd><p>List of k-mer sizes</p>
</dd>
<dt>raw_matching (list)</dt><dd><p>Proportion of matching k-mers at each klist value</p>
</dd>
<dt>raw_fit (numpy.array)</dt><dd><p>Fit to klist and raw_matching from <a class="reference internal" href="#PopPUNK.sketchlib.fitKmerCurve" title="PopPUNK.sketchlib.fitKmerCurve"><code class="xref py py-func docutils literal notranslate"><span class="pre">fitKmerCurve()</span></code></a></p>
</dd>
<dt>corrected_matching (list)</dt><dd><p>Corrected proportion of matching k-mers at each klist value</p>
</dd>
<dt>corrected_fit (numpy.array)</dt><dd><p>Fit to klist and corrected_matching from <a class="reference internal" href="#PopPUNK.sketchlib.fitKmerCurve" title="PopPUNK.sketchlib.fitKmerCurve"><code class="xref py py-func docutils literal notranslate"><span class="pre">fitKmerCurve()</span></code></a></p>
</dd>
<dt>out_prefix (str)</dt><dd><p>Prefix for output plot file (.pdf will be appended)</p>
</dd>
<dt>title (str)</dt><dd><p>The title to display above the plot</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.plot_refined_results">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">plot_refined_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_boundary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_boundary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">core_boundary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accessory_boundary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_move</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_move</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indiv_boundaries</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unconstrained</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_prefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#plot_refined_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.plot_refined_results" title="Permalink to this definition">#</a></dt>
<dd><p>Draw a scatter plot (png) to show the refined model fit</p>
<p>A scatter plot of core and accessory distances, coloured by component
membership. The triangular decision boundary is also shown</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>n x 2 array of core and accessory distances for n samples.</p>
</dd>
<dt>Y (numpy.array)</dt><dd><p>n x 1 array of cluster assignments for n samples.</p>
</dd>
<dt>x_boundary (float)</dt><dd><p>Intercept of boundary with x-axis, from <a class="reference internal" href="#PopPUNK.models.RefineFit" title="PopPUNK.models.RefineFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">RefineFit</span></code></a></p>
</dd>
<dt>y_boundary (float)</dt><dd><p>Intercept of boundary with y-axis, from <a class="reference internal" href="#PopPUNK.models.RefineFit" title="PopPUNK.models.RefineFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">RefineFit</span></code></a></p>
</dd>
<dt>core_boundary (float)</dt><dd><p>Intercept of 1D (core) boundary with x-axis, from <a class="reference internal" href="#PopPUNK.models.RefineFit" title="PopPUNK.models.RefineFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">RefineFit</span></code></a></p>
</dd>
<dt>accessory_boundary (float)</dt><dd><p>Intercept of 1D (core) boundary with y-axis, from <a class="reference internal" href="#PopPUNK.models.RefineFit" title="PopPUNK.models.RefineFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">RefineFit</span></code></a></p>
</dd>
<dt>mean0 (numpy.array)</dt><dd><p>Centre of within-strain distribution</p>
</dd>
<dt>mean1 (numpy.array)</dt><dd><p>Centre of between-strain distribution</p>
</dd>
<dt>min_move (float)</dt><dd><p>Minimum s range</p>
</dd>
<dt>max_move (float)</dt><dd><p>Maximum s range</p>
</dd>
<dt>scale (numpy.array)</dt><dd><p>Scaling factor from <a class="reference internal" href="#PopPUNK.models.RefineFit" title="PopPUNK.models.RefineFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">RefineFit</span></code></a></p>
</dd>
<dt>threshold (bool)</dt><dd><p>If fit was just from a simple thresholding</p>
</dd>
<dt>indiv_boundaries (bool)</dt><dd><p>Whether to draw lines for core and accessory refinement</p>
</dd>
<dt>title (str)</dt><dd><p>The title to display above the plot</p>
</dd>
<dt>out_prefix (str)</dt><dd><p>Prefix for output plot file (.png will be appended)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.plot_results">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">plot_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">means</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">covariances</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_prefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#plot_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.plot_results" title="Permalink to this definition">#</a></dt>
<dd><p>Draw a scatter plot (png) to show the BGMM model fit</p>
<p>A scatter plot of core and accessory distances, coloured by component
membership. Also shown are ellipses for each component (centre: means
axes: covariances).</p>
<p>This is based on the example in the sklearn documentation.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X (numpy.array)</dt><dd><p>n x 2 array of core and accessory distances for n samples.</p>
</dd>
<dt>Y (numpy.array)</dt><dd><p>n x 1 array of cluster assignments for n samples.</p>
</dd>
<dt>means (numpy.array)</dt><dd><p>Component means from <a class="reference internal" href="#PopPUNK.models.BGMMFit" title="PopPUNK.models.BGMMFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">BGMMFit</span></code></a></p>
</dd>
<dt>covars (numpy.array)</dt><dd><p>Component covariances from <a class="reference internal" href="#PopPUNK.models.BGMMFit" title="PopPUNK.models.BGMMFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">BGMMFit</span></code></a></p>
</dd>
<dt>scale (numpy.array)</dt><dd><p>Scaling factor from <a class="reference internal" href="#PopPUNK.models.BGMMFit" title="PopPUNK.models.BGMMFit"><code class="xref py py-class docutils literal notranslate"><span class="pre">BGMMFit</span></code></a></p>
</dd>
<dt>out_prefix (str)</dt><dd><p>Prefix for output plot file (.png will be appended)</p>
</dd>
<dt>title (str)</dt><dd><p>The title to display above the plot</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.plot_scatter">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">plot_scatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kde</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#plot_scatter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.plot_scatter" title="Permalink to this definition">#</a></dt>
<dd><p>Draws a 2D scatter plot (png) of the core and accessory distances</p>
<p>Also draws contours of kernel density estimare</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>X (numpy.array)</dt><dd><p>n x 2 array of core and accessory distances for n samples.</p>
</dd>
<dt>out_prefix (str)</dt><dd><p>Prefix for output plot file (.png will be appended)</p>
</dd>
<dt>title (str)</dt><dd><p>The title to display above the plot</p>
</dd>
<dt>kde (bool)</dt><dd><p>Whether to draw kernel density estimate contours</p>
<p>(default = True)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.plot.writeClusterCsv">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.plot.</span></span><span class="sig-name descname"><span class="pre">writeClusterCsv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodeNames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodeLabels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'microreact'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epiCsv</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">queryNames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_Cluster'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/plot.html#writeClusterCsv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.plot.writeClusterCsv" title="Permalink to this definition">#</a></dt>
<dd><p>Print CSV file of clustering and optionally epi data</p>
<p>Writes CSV output of clusters which can be used as input to microreact and cytoscape.
Uses pandas to deal with CSV reading and writing nicely.</p>
<p>The epiCsv, if provided, should have the node labels in the first column.</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>outfile (str)</dt><dd><p>File to write the CSV to.</p>
</dd>
<dt>nodeNames (list)</dt><dd><p>Names of sequences in clustering (includes path).</p>
</dd>
<dt>nodeLabels (list)</dt><dd><p>Names of sequences to write in CSV (usually has path removed).</p>
</dd>
<dt>clustering (dict or dict of dicts)</dt><dd><p>Dictionary of cluster assignments (keys are nodeNames). Pass a dict with depth two
to include multiple possible clusterings.</p>
</dd>
<dt>output_format (str)</dt><dd><p>Software for which CSV should be formatted
(microreact, phandango, grapetree and cytoscape are accepted)</p>
</dd>
<dt>epiCsv (str)</dt><dd><p>Optional CSV of epi data to paste in the output in addition to
the clusters (default = None).</p>
</dd>
<dt>queryNames (list)</dt><dd><p>Optional list of isolates that have been added as a query.</p>
<p>(default = None)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-PopPUNK.sparse_mst">
<span id="sparse-mst-py"></span><h2>sparse_mst.py<a class="headerlink" href="#module-PopPUNK.sparse_mst" title="Permalink to this heading">#</a></h2>
</section>
<section id="module-PopPUNK.sketchlib">
<span id="sketchlib-py"></span><h2>sketchlib.py<a class="headerlink" href="#module-PopPUNK.sketchlib" title="Permalink to this heading">#</a></h2>
<p>Sketchlib functions for database construction</p>
<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.addRandom">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">addRandom</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">oPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">klist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_preserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#addRandom"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.addRandom" title="Permalink to this definition">#</a></dt>
<dd><p>Add chance of random match to a HDF5 sketch DB</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>oPrefix (str)</dt><dd><p>Sketch database prefix</p>
</dd>
<dt>sequence_names (list)</dt><dd><p>Names of sequences to include in calculation</p>
</dd>
<dt>klist (list)</dt><dd><p>List of k-mer sizes to sketch</p>
</dd>
<dt>strand_preserved (bool)</dt><dd><p>Set true to ignore rc k-mers</p>
</dd>
<dt>overwrite (str)</dt><dd><p>Set true to overwrite existing random match chances</p>
</dd>
<dt>threads (int)</dt><dd><p>Number of threads to use (default = 1)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.checkSketchlibLibrary">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">checkSketchlibLibrary</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#checkSketchlibLibrary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.checkSketchlibLibrary" title="Permalink to this definition">#</a></dt>
<dd><p>Gets the location of the sketchlib library</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>lib (str)</dt><dd><p>Location of sketchlib .so/.dyld</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.checkSketchlibVersion">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">checkSketchlibVersion</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#checkSketchlibVersion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.checkSketchlibVersion" title="Permalink to this definition">#</a></dt>
<dd><p>Checks that sketchlib can be run, and returns version</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>version (str)</dt><dd><p>Version string</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.constructDatabase">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">constructDatabase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assemblyList</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">klist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sketch_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">oPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_preserved</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_count</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_exact</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calc_random</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">codon_phased</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">deviceid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#constructDatabase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.constructDatabase" title="Permalink to this definition">#</a></dt>
<dd><p>Sketch the input assemblies at the requested k-mer lengths</p>
<p>A multithread wrapper around <code class="xref py py-func docutils literal notranslate"><span class="pre">runSketch()</span></code>. Threads are used to either run multiple sketch
processes for each klist value.</p>
<p>Also calculates random match probability based on length of first genome
in assemblyList.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>assemblyList (str)</dt><dd><p>File with locations of assembly files to be sketched</p>
</dd>
<dt>klist (list)</dt><dd><p>List of k-mer sizes to sketch</p>
</dd>
<dt>sketch_size (int)</dt><dd><p>Size of sketch (<code class="docutils literal notranslate"><span class="pre">-s</span></code> option)</p>
</dd>
<dt>oPrefix (str)</dt><dd><p>Output prefix for resulting sketch files</p>
</dd>
<dt>threads (int)</dt><dd><p>Number of threads to use (default = 1)</p>
</dd>
<dt>overwrite (bool)</dt><dd><p>Whether to overwrite sketch DBs, if they already exist.
(default = False)</p>
</dd>
<dt>strand_preserved (bool)</dt><dd><p>Ignore reverse complement k-mers (default = False)</p>
</dd>
<dt>min_count (int)</dt><dd><p>Minimum count of k-mer in reads to include
(default = 0)</p>
</dd>
<dt>use_exact (bool)</dt><dd><p>Use exact count of k-mer appearance in reads
(default = False)</p>
</dd>
<dt>calc_random (bool)</dt><dd><p>Add random match chances to DB (turn off for queries)</p>
</dd>
<dt>codon_phased (bool)</dt><dd><p>Use codon phased seeds
(default = False)</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Use GPU for read sketching
(default = False)</p>
</dd>
<dt>deviceid (int)</dt><dd><p>GPU device id
(default = 0)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>names (list)</dt><dd><p>List of names included in the database (from rfile)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.createDatabaseDir">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">createDatabaseDir</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kmers</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#createDatabaseDir"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.createDatabaseDir" title="Permalink to this definition">#</a></dt>
<dd><p>Creates the directory to write sketches to, removing old files if unnecessary</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>outPrefix (str)</dt><dd><p>output db prefix</p>
</dd>
<dt>kmers (list)</dt><dd><p>k-mer sizes in db</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.fitKmerCurve">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">fitKmerCurve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pairwise</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">klist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jacobian</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#fitKmerCurve"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.fitKmerCurve" title="Permalink to this definition">#</a></dt>
<dd><p>Fit the function <span class="math notranslate nohighlight">\(pr = (1-a)(1-c)^k\)</span></p>
<p>Supply <code class="docutils literal notranslate"><span class="pre">jacobian</span> <span class="pre">=</span> <span class="pre">-np.hstack((np.ones((klist.shape[0],</span> <span class="pre">1)),</span> <span class="pre">klist.reshape(-1,</span> <span class="pre">1)))</span></code></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>pairwise (numpy.array)</dt><dd><p>Proportion of shared k-mers at k-mer values in klist</p>
</dd>
<dt>klist (list)</dt><dd><p>k-mer sizes used</p>
</dd>
<dt>jacobian (numpy.array)</dt><dd><p>Should be set as above (set once to try and save memory)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>transformed_params (numpy.array)</dt><dd><p>Column with core and accessory distance</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.getKmersFromReferenceDatabase">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">getKmersFromReferenceDatabase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbPrefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#getKmersFromReferenceDatabase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.getKmersFromReferenceDatabase" title="Permalink to this definition">#</a></dt>
<dd><p>Get kmers lengths from existing database</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>dbPrefix (str)</dt><dd><p>Prefix for sketch DB files</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>kmers (list)</dt><dd><p>List of k-mer lengths used in database</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.getSeqsInDb">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">getSeqsInDb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbname</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#getSeqsInDb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.getSeqsInDb" title="Permalink to this definition">#</a></dt>
<dd><p>Return an array with the sequences in the passed database</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>dbname (str)</dt><dd><p>Sketches database filename</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>seqs (list)</dt><dd><p>List of sequence names in sketch DB</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.getSketchSize">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">getSketchSize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbPrefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#getSketchSize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.getSketchSize" title="Permalink to this definition">#</a></dt>
<dd><p>Determine sketch size, and ensures consistent in whole database</p>
<p><code class="docutils literal notranslate"><span class="pre">sys.exit(1)</span></code> is called if DBs have different sketch sizes</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>dbprefix (str)</dt><dd><p>Prefix for databases</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>sketchSize (int)</dt><dd><p>sketch size (64x C++ definition)</p>
</dd>
<dt>codonPhased (bool)</dt><dd><p>whether the DB used codon phased seeds</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.joinDBs">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">joinDBs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update_random</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#joinDBs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.joinDBs" title="Permalink to this definition">#</a></dt>
<dd><p>Join two sketch databases with the low-level HDF5 copy interface</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>db1 (str)</dt><dd><p>Prefix for db1</p>
</dd>
<dt>db2 (str)</dt><dd><p>Prefix for db2</p>
</dd>
<dt>output (str)</dt><dd><p>Prefix for joined output</p>
</dd>
<dt>update_random (dict)</dt><dd><p>Whether to re-calculate the random object. May contain
control arguments strand_preserved and threads (see <a class="reference internal" href="#PopPUNK.sketchlib.addRandom" title="PopPUNK.sketchlib.addRandom"><code class="xref py py-func docutils literal notranslate"><span class="pre">addRandom()</span></code></a>)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.queryDatabase">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">queryDatabase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rNames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qNames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dbPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">queryPrefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">klist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">self</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">number_plot_fits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">deviceid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#queryDatabase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.queryDatabase" title="Permalink to this definition">#</a></dt>
<dd><p>Calculate core and accessory distances between query sequences and a sketched database</p>
<p>For a reference database, runs the query against itself to find all pairwise
core and accessory distances.</p>
<p>Uses the relation <span class="math notranslate nohighlight">\(pr(a, b) = (1-a)(1-c)^k\)</span></p>
<p>To get the ref and query name for each row of the returned distances, call to the iterator
<a class="reference internal" href="#PopPUNK.utils.iterDistRows" title="PopPUNK.utils.iterDistRows"><code class="xref py py-func docutils literal notranslate"><span class="pre">iterDistRows()</span></code></a> with the returned refList and queryList</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>rNames (list)</dt><dd><p>Names of references to query</p>
</dd>
<dt>qNames (list)</dt><dd><p>Names of queries</p>
</dd>
<dt>dbPrefix (str)</dt><dd><p>Prefix for reference sketch database created by <a class="reference internal" href="#PopPUNK.sketchlib.constructDatabase" title="PopPUNK.sketchlib.constructDatabase"><code class="xref py py-func docutils literal notranslate"><span class="pre">constructDatabase()</span></code></a></p>
</dd>
<dt>queryPrefix (str)</dt><dd><p>Prefix for query sketch database created by <a class="reference internal" href="#PopPUNK.sketchlib.constructDatabase" title="PopPUNK.sketchlib.constructDatabase"><code class="xref py py-func docutils literal notranslate"><span class="pre">constructDatabase()</span></code></a></p>
</dd>
<dt>klist (list)</dt><dd><p>K-mer sizes to use in the calculation</p>
</dd>
<dt>self (bool)</dt><dd><p>Set true if query = ref
(default = True)</p>
</dd>
<dt>number_plot_fits (int)</dt><dd><p>If &gt; 0, the number of k-mer length fits to plot (saved as pdfs).
Takes random pairs of comparisons and calls <a class="reference internal" href="#PopPUNK.plot.plot_fit" title="PopPUNK.plot.plot_fit"><code class="xref py py-func docutils literal notranslate"><span class="pre">plot_fit()</span></code></a>
(default = 0)</p>
</dd>
<dt>threads (int)</dt><dd><p>Number of threads to use in the process
(default = 1)</p>
</dd>
<dt>use_gpu (bool)</dt><dd><p>Use a GPU for querying
(default = False)</p>
</dd>
<dt>deviceid (int)</dt><dd><p>Index of the CUDA GPU device to use
(default = 0)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>distMat (numpy.array)</dt><dd><p>Core distances (column 0) and accessory distances (column 1) between
refList and queryList</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.readDBParams">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">readDBParams</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbPrefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#readDBParams"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.readDBParams" title="Permalink to this definition">#</a></dt>
<dd><p>Get kmers lengths and sketch sizes from existing database</p>
<p>Calls <a class="reference internal" href="#PopPUNK.sketchlib.getKmersFromReferenceDatabase" title="PopPUNK.sketchlib.getKmersFromReferenceDatabase"><code class="xref py py-func docutils literal notranslate"><span class="pre">getKmersFromReferenceDatabase()</span></code></a> and <a class="reference internal" href="#PopPUNK.sketchlib.getSketchSize" title="PopPUNK.sketchlib.getSketchSize"><code class="xref py py-func docutils literal notranslate"><span class="pre">getSketchSize()</span></code></a>
Uses passed values if db missing</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>dbPrefix (str)</dt><dd><p>Prefix for sketch DB files</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>kmers (list)</dt><dd><p>List of k-mer lengths used in database</p>
</dd>
<dt>sketch_sizes (list)</dt><dd><p>List of sketch sizes used in database</p>
</dd>
<dt>codonPhased (bool)</dt><dd><p>whether the DB used codon phased seeds</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.sketchlib.removeFromDB">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.sketchlib.</span></span><span class="sig-name descname"><span class="pre">removeFromDB</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">removeSeqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/sketchlib.html#removeFromDB"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.sketchlib.removeFromDB" title="Permalink to this definition">#</a></dt>
<dd><p>Remove sketches from the DB the low-level HDF5 copy interface</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>db_name (str)</dt><dd><p>Prefix for hdf database</p>
</dd>
<dt>out_name (str)</dt><dd><p>Prefix for output (pruned) database</p>
</dd>
<dt>removeSeqs (list)</dt><dd><p>Names of sequences to remove from database</p>
</dd>
<dt>full_names (bool)</dt><dd><p>If True, db_name and out_name are the full paths to h5 files</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-PopPUNK.utils">
<span id="utils-py"></span><h2>utils.py<a class="headerlink" href="#module-PopPUNK.utils" title="Permalink to this heading">#</a></h2>
<p>General utility functions for data read/writing/manipulation in PopPUNK</p>
<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.check_and_set_gpu">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">check_and_set_gpu</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gpu_lib</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quit_on_fail</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#check_and_set_gpu"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.check_and_set_gpu" title="Permalink to this definition">#</a></dt>
<dd><p>Check GPU libraries can be loaded and set managed memory.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>use_gpu (bool)</dt><dd><p>Whether GPU packages have been requested</p>
</dd>
<dt>gpu_lib (bool)</dt><dd><p>Whether GPU packages are available</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>use_gpu (bool)</dt><dd><p>Whether GPU packages can be used</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.decisionBoundary">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">decisionBoundary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intercept</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gradient</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#decisionBoundary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.decisionBoundary" title="Permalink to this definition">#</a></dt>
<dd><p>Returns the co-ordinates where the triangle the decision boundary forms
meets the x- and y-axes.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>intercept (numpy.array)</dt><dd><p>Cartesian co-ordinates of point along line (<a class="reference internal" href="#PopPUNK.utils.transformLine" title="PopPUNK.utils.transformLine"><code class="xref py py-func docutils literal notranslate"><span class="pre">transformLine()</span></code></a>)
which intercepts the boundary</p>
</dd>
<dt>gradient (float)</dt><dd><p>Gradient of the line</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>x (float)</dt><dd><p>The x-axis intercept</p>
</dd>
<dt>y (float)</dt><dd><p>The y-axis intercept</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.get_match_search_depth">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">get_match_search_depth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#get_match_search_depth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.get_match_search_depth" title="Permalink to this definition">#</a></dt>
<dd><p>Return a default search depth for lineage model fitting.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>rlist (list)</dt><dd><p>List of sequences in database</p>
</dd>
<dt>rank_list (list)</dt><dd><p>List of ranks to be used to fit lineage models</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>max_search_depth (int)</dt><dd><p>Maximum kNN used for lineage model fitting</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.isolateNameToLabel">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">isolateNameToLabel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">names</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#isolateNameToLabel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.isolateNameToLabel" title="Permalink to this definition">#</a></dt>
<dd><p>Function to process isolate names to labels
appropriate for visualisation.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>names (list)</dt><dd><p>List of isolate names.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>labels (list)</dt><dd><p>List of isolate labels.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.iterDistRows">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">iterDistRows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">refSeqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">querySeqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">self</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#iterDistRows"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.iterDistRows" title="Permalink to this definition">#</a></dt>
<dd><p>Gets the ref and query ID for each row of the distance matrix</p>
<p>Returns an iterable with ref and query ID pairs by row.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>refSeqs (list)</dt><dd><p>List of reference sequence names.</p>
</dd>
<dt>querySeqs (list)</dt><dd><p>List of query sequence names.</p>
</dd>
<dt>self (bool)</dt><dd><p>Whether a self-comparison, used when constructing a database.
Requires refSeqs == querySeqs
Default is True</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>ref, query (str, str)</dt><dd><p>Iterable of tuples with ref and query names for each distMat row.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.joinClusterDicts">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">joinClusterDicts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#joinClusterDicts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.joinClusterDicts" title="Permalink to this definition">#</a></dt>
<dd><p>Join two dictionaries returned by <a class="reference internal" href="#PopPUNK.utils.readIsolateTypeFromCsv" title="PopPUNK.utils.readIsolateTypeFromCsv"><code class="xref py py-func docutils literal notranslate"><span class="pre">readIsolateTypeFromCsv()</span></code></a> with
return_dict = True. Useful for concatenating ref and query assignments</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>d1 (dict of dicts)</dt><dd><p>First dictionary to concat</p>
</dd>
<dt>d2 (dict of dicts)</dt><dd><p>Second dictionary to concat</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>d1 (dict of dicts)</dt><dd><p>d1 with d2 appended</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.listDistInts">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">listDistInts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">refSeqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">querySeqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">self</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#listDistInts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.listDistInts" title="Permalink to this definition">#</a></dt>
<dd><p>Gets the ref and query ID for each row of the distance matrix</p>
<p>Returns an iterable with ref and query ID pairs by row.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>refSeqs (list)</dt><dd><p>List of reference sequence names.</p>
</dd>
<dt>querySeqs (list)</dt><dd><p>List of query sequence names.</p>
</dd>
<dt>self (bool)</dt><dd><p>Whether a self-comparison, used when constructing a database.
Requires refSeqs == querySeqs
Default is True</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>ref, query (str, str)</dt><dd><p>Iterable of tuples with ref and query names for each distMat row.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.readIsolateTypeFromCsv">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">readIsolateTypeFromCsv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">clustCSV</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#readIsolateTypeFromCsv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.readIsolateTypeFromCsv" title="Permalink to this definition">#</a></dt>
<dd><p>Read cluster definitions from CSV file.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>clustCSV (str)</dt><dd><p>File name of CSV with isolate assignments</p>
</dd>
<dt>mode (str)</dt><dd><p>Type of file to read ‘clusters’, ‘lineages’, or ‘external’</p>
</dd>
<dt>return_type (str)</dt><dd><p>If True, return a dict with sample-&gt;cluster instead
of sets
[default = False]</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>clusters (dict)</dt><dd><p>Dictionary of cluster assignments (keys are cluster names, values are
sets containing samples in the cluster). Or if return_dict is set keys
are sample names, values are cluster assignments.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.readPickle">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">readPickle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pklName</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enforce_self</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#readPickle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.readPickle" title="Permalink to this definition">#</a></dt>
<dd><p>Loads core and accessory distances saved by <a class="reference internal" href="#PopPUNK.utils.storePickle" title="PopPUNK.utils.storePickle"><code class="xref py py-func docutils literal notranslate"><span class="pre">storePickle()</span></code></a></p>
<p>Called during <code class="docutils literal notranslate"><span class="pre">--fit-model</span></code></p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>pklName (str)</dt><dd><p>Prefix for saved files</p>
</dd>
<dt>enforce_self (bool)</dt><dd><p>Error if self == False</p>
<p>[default = True]</p>
</dd>
<dt>distances (bool)</dt><dd><p>Read the distance matrix</p>
<p>[default = True]</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>rlist (list)</dt><dd><p>List of reference sequence names (for <a class="reference internal" href="#PopPUNK.utils.iterDistRows" title="PopPUNK.utils.iterDistRows"><code class="xref py py-func docutils literal notranslate"><span class="pre">iterDistRows()</span></code></a>)</p>
</dd>
<dt>qlist (list)</dt><dd><p>List of query sequence names (for <a class="reference internal" href="#PopPUNK.utils.iterDistRows" title="PopPUNK.utils.iterDistRows"><code class="xref py py-func docutils literal notranslate"><span class="pre">iterDistRows()</span></code></a>)</p>
</dd>
<dt>self (bool)</dt><dd><p>Whether an all-vs-all self DB (for <a class="reference internal" href="#PopPUNK.utils.iterDistRows" title="PopPUNK.utils.iterDistRows"><code class="xref py py-func docutils literal notranslate"><span class="pre">iterDistRows()</span></code></a>)</p>
</dd>
<dt>X (numpy.array)</dt><dd><p>n x 2 array of core and accessory distances</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.readRfile">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">readRfile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rFile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">oneSeq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#readRfile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.readRfile" title="Permalink to this definition">#</a></dt>
<dd><p>Reads in files for sketching. Names and sequence, tab separated</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>rFile (str)</dt><dd><p>File with locations of assembly files to be sketched</p>
</dd>
<dt>oneSeq (bool)</dt><dd><p>Return only the first sequence listed, rather than a list
(used with mash)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>names (list)</dt><dd><p>Array of sequence names</p>
</dd>
<dt>sequences (list of lists)</dt><dd><p>Array of sequence files</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.read_rlist_from_distance_pickle">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">read_rlist_from_distance_pickle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_non_self</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#read_rlist_from_distance_pickle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.read_rlist_from_distance_pickle" title="Permalink to this definition">#</a></dt>
<dd><p>Return the list of reference sequences from a distance pickle.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>fn (str)</dt><dd><p>Name of distance pickle</p>
</dd>
<dt>allow_non_self (bool)</dt><dd><p>Whether non-self distance datasets are permissible</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>rlist (list)</dt><dd><p>List of reference sequence names</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.set_env">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">set_env</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">environ</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#set_env"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.set_env" title="Permalink to this definition">#</a></dt>
<dd><p>Temporarily set the process environment variables.
&gt;&gt;&gt; with set_env(PLUGINS_DIR=u’test/plugins’):
…   “PLUGINS_DIR” in os.environ
True
&gt;&gt;&gt; “PLUGINS_DIR” in os.environ
False</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.setupDBFuncs">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">setupDBFuncs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#setupDBFuncs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.setupDBFuncs" title="Permalink to this definition">#</a></dt>
<dd><p>Wraps common database access functions from sketchlib and mash,
to try and make their API more similar</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>args (argparse.opts)</dt><dd><p>Parsed command lines options</p>
</dd>
<dt>qc_dict (dict)</dt><dd><p>Table of parameters for QC function</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>dbFuncs (dict)</dt><dd><p>Functions with consistent arguments to use as the database API</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.stderr_redirected">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">stderr_redirected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/dev/null'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#stderr_redirected"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.stderr_redirected" title="Permalink to this definition">#</a></dt>
<dd><p>import os</p>
<dl class="simple">
<dt>with stdout_redirected(to=filename):</dt><dd><p>print(“from Python”)
os.system(“echo non-Python applications are also supported”)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.storePickle">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">storePickle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qlist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pklName</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#storePickle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.storePickle" title="Permalink to this definition">#</a></dt>
<dd><p>Saves core and accessory distances in a .npy file, names in a .pkl</p>
<p>Called during <code class="docutils literal notranslate"><span class="pre">--create-db</span></code></p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>rlist (list)</dt><dd><p>List of reference sequence names (for <a class="reference internal" href="#PopPUNK.utils.iterDistRows" title="PopPUNK.utils.iterDistRows"><code class="xref py py-func docutils literal notranslate"><span class="pre">iterDistRows()</span></code></a>)</p>
</dd>
<dt>qlist (list)</dt><dd><p>List of query sequence names (for <a class="reference internal" href="#PopPUNK.utils.iterDistRows" title="PopPUNK.utils.iterDistRows"><code class="xref py py-func docutils literal notranslate"><span class="pre">iterDistRows()</span></code></a>)</p>
</dd>
<dt>self (bool)</dt><dd><p>Whether an all-vs-all self DB (for <a class="reference internal" href="#PopPUNK.utils.iterDistRows" title="PopPUNK.utils.iterDistRows"><code class="xref py py-func docutils literal notranslate"><span class="pre">iterDistRows()</span></code></a>)</p>
</dd>
<dt>X (numpy.array)</dt><dd><p>n x 2 array of core and accessory distances</p>
</dd>
<dt>pklName (str)</dt><dd><p>Prefix for output files</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.transformLine">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">transformLine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#transformLine"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.transformLine" title="Permalink to this definition">#</a></dt>
<dd><p>Return x and y co-ordinates for traversing along a line between mean0 and mean1, parameterised by
a single scalar distance s from the start point mean0.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>s (float)</dt><dd><p>Distance along line from mean0</p>
</dd>
<dt>mean0 (numpy.array)</dt><dd><p>Start position of line (x0, y0)</p>
</dd>
<dt>mean1 (numpy.array)</dt><dd><p>End position of line (x1, y1)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>x (float)</dt><dd><p>The Cartesian x-coordinate</p>
</dd>
<dt>y (float)</dt><dd><p>The Cartesian y-coordinate</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.utils.update_distance_matrices">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.utils.</span></span><span class="sig-name descname"><span class="pre">update_distance_matrices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">refList</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distMat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">queryList</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query_ref_distMat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query_query_distMat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/utils.html#update_distance_matrices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.utils.update_distance_matrices" title="Permalink to this definition">#</a></dt>
<dd><p>Convert distances from long form (1 matrix with n_comparisons rows and 2 columns)
to a square form (2 NxN matrices), with merging of query distances if necessary.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>refList (list)</dt><dd><p>List of references</p>
</dd>
<dt>distMat (numpy.array)</dt><dd><p>Two column long form list of core and accessory distances
for pairwise comparisons between reference db sequences</p>
</dd>
<dt>queryList (list)</dt><dd><p>List of queries</p>
</dd>
<dt>query_ref_distMat (numpy.array)</dt><dd><p>Two column long form list of core and accessory distances
for pairwise comparisons between queries and reference db
sequences</p>
</dd>
<dt>query_query_distMat (numpy.array)</dt><dd><p>Two column long form list of core and accessory distances
for pairwise comparisons between query sequences</p>
</dd>
<dt>threads (int)</dt><dd><p>Number of threads to use</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>seqLabels (list)</dt><dd><p>Combined list of reference and query sequences</p>
</dd>
<dt>coreMat (numpy.array)</dt><dd><p>NxN array of core distances for N sequences</p>
</dd>
<dt>accMat (numpy.array)</dt><dd><p>NxN array of accessory distances for N sequences</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="visualise-py">
<h2>visualise.py<a class="headerlink" href="#visualise-py" title="Permalink to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">poppunk_visualise</span></code> main function</p>
<span class="target" id="module-PopPUNK.visualise"></span><dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.visualise.main">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.visualise.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/visualise.html#main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.visualise.main" title="Permalink to this definition">#</a></dt>
<dd><p>Main function. Parses cmd line args and runs in the specified mode.</p>
</dd></dl>

</section>
<section id="web-py">
<h2>web.py<a class="headerlink" href="#web-py" title="Permalink to this heading">#</a></h2>
<p>Functions used by the web API to convert a sketch to an h5 database, then generate visualisations and post results to PopPUNK-web.</p>
<span class="target" id="module-PopPUNK.web"></span><dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.web.api">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.web.</span></span><span class="sig-name descname"><span class="pre">api</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_db</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/web.html#api"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.web.api" title="Permalink to this definition">#</a></dt>
<dd><p>Post cluster and tree information to microreact</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.web.calc_prevalence">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.web.</span></span><span class="sig-name descname"><span class="pre">calc_prevalence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cluster</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_samples</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/web.html#calc_prevalence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.web.calc_prevalence" title="Permalink to this definition">#</a></dt>
<dd><p>Cluster prevalences for Plotly.js</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.web.graphml_to_json">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.web.</span></span><span class="sig-name descname"><span class="pre">graphml_to_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_dir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/web.html#graphml_to_json"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.web.graphml_to_json" title="Permalink to this definition">#</a></dt>
<dd><p>Converts full GraphML file to JSON subgraph</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.web.highlight_cluster">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.web.</span></span><span class="sig-name descname"><span class="pre">highlight_cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/web.html#highlight_cluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.web.highlight_cluster" title="Permalink to this definition">#</a></dt>
<dd><p>Colour assigned cluster in Microreact output</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.web.sketch_to_hdf5">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.web.</span></span><span class="sig-name descname"><span class="pre">sketch_to_hdf5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sketches_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/web.html#sketch_to_hdf5"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.web.sketch_to_hdf5" title="Permalink to this definition">#</a></dt>
<dd><p>Convert dict of JSON sketches to query hdf5 database</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="PopPUNK.web.summarise_clusters">
<span class="sig-prename descclassname"><span class="pre">PopPUNK.web.</span></span><span class="sig-name descname"><span class="pre">summarise_clusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species_db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qNames</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/PopPUNK/web.html#summarise_clusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PopPUNK.web.summarise_clusters" title="Permalink to this definition">#</a></dt>
<dd><p>Retreieve assigned query and all cluster prevalences.
Write list of all isolates in cluster for tree subsetting</p>
</dd></dl>

</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="roadmap.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Roadmap</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="citing.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Citing PopPUNK</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2018-2023, John Lees and Nicholas Croucher
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Reference documentation</a><ul>
<li><a class="reference internal" href="#assign-py">assign.py</a><ul>
<li><a class="reference internal" href="#PopPUNK.assign.assign_query"><code class="docutils literal notranslate"><span class="pre">assign_query()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.assign.assign_query_hdf5"><code class="docutils literal notranslate"><span class="pre">assign_query_hdf5()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.assign.main"><code class="docutils literal notranslate"><span class="pre">main()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#bgmm-py">bgmm.py</a><ul>
<li><a class="reference internal" href="#PopPUNK.bgmm.findWithinLabel"><code class="docutils literal notranslate"><span class="pre">findWithinLabel()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.bgmm.fit2dMultiGaussian"><code class="docutils literal notranslate"><span class="pre">fit2dMultiGaussian()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.bgmm.log_likelihood"><code class="docutils literal notranslate"><span class="pre">log_likelihood()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.bgmm.log_multivariate_normal_density"><code class="docutils literal notranslate"><span class="pre">log_multivariate_normal_density()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#dbscan-py">dbscan.py</a><ul>
<li><a class="reference internal" href="#PopPUNK.dbscan.evaluate_dbscan_clusters"><code class="docutils literal notranslate"><span class="pre">evaluate_dbscan_clusters()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.dbscan.findBetweenLabel"><code class="docutils literal notranslate"><span class="pre">findBetweenLabel()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.dbscan.fitDbScan"><code class="docutils literal notranslate"><span class="pre">fitDbScan()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-PopPUNK.mandrake">mandrake.py</a><ul>
<li><a class="reference internal" href="#PopPUNK.mandrake.generate_embedding"><code class="docutils literal notranslate"><span class="pre">generate_embedding()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-PopPUNK.models">models.py</a><ul>
<li><a class="reference internal" href="#PopPUNK.models.BGMMFit"><code class="docutils literal notranslate"><span class="pre">BGMMFit</span></code></a><ul>
<li><a class="reference internal" href="#PopPUNK.models.BGMMFit.assign"><code class="docutils literal notranslate"><span class="pre">BGMMFit.assign()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.BGMMFit.fit"><code class="docutils literal notranslate"><span class="pre">BGMMFit.fit()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.BGMMFit.load"><code class="docutils literal notranslate"><span class="pre">BGMMFit.load()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.BGMMFit.plot"><code class="docutils literal notranslate"><span class="pre">BGMMFit.plot()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.BGMMFit.save"><code class="docutils literal notranslate"><span class="pre">BGMMFit.save()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#PopPUNK.models.ClusterFit"><code class="docutils literal notranslate"><span class="pre">ClusterFit</span></code></a><ul>
<li><a class="reference internal" href="#PopPUNK.models.ClusterFit.copy"><code class="docutils literal notranslate"><span class="pre">ClusterFit.copy()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.ClusterFit.fit"><code class="docutils literal notranslate"><span class="pre">ClusterFit.fit()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.ClusterFit.no_scale"><code class="docutils literal notranslate"><span class="pre">ClusterFit.no_scale()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.ClusterFit.plot"><code class="docutils literal notranslate"><span class="pre">ClusterFit.plot()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#PopPUNK.models.DBSCANFit"><code class="docutils literal notranslate"><span class="pre">DBSCANFit</span></code></a><ul>
<li><a class="reference internal" href="#PopPUNK.models.DBSCANFit.assign"><code class="docutils literal notranslate"><span class="pre">DBSCANFit.assign()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.DBSCANFit.fit"><code class="docutils literal notranslate"><span class="pre">DBSCANFit.fit()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.DBSCANFit.load"><code class="docutils literal notranslate"><span class="pre">DBSCANFit.load()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.DBSCANFit.plot"><code class="docutils literal notranslate"><span class="pre">DBSCANFit.plot()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.DBSCANFit.save"><code class="docutils literal notranslate"><span class="pre">DBSCANFit.save()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#PopPUNK.models.LineageFit"><code class="docutils literal notranslate"><span class="pre">LineageFit</span></code></a><ul>
<li><a class="reference internal" href="#PopPUNK.models.LineageFit.assign"><code class="docutils literal notranslate"><span class="pre">LineageFit.assign()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.LineageFit.edge_weights"><code class="docutils literal notranslate"><span class="pre">LineageFit.edge_weights()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.LineageFit.extend"><code class="docutils literal notranslate"><span class="pre">LineageFit.extend()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.LineageFit.fit"><code class="docutils literal notranslate"><span class="pre">LineageFit.fit()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.LineageFit.load"><code class="docutils literal notranslate"><span class="pre">LineageFit.load()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.LineageFit.plot"><code class="docutils literal notranslate"><span class="pre">LineageFit.plot()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.LineageFit.save"><code class="docutils literal notranslate"><span class="pre">LineageFit.save()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#PopPUNK.models.NumpyShared"><code class="docutils literal notranslate"><span class="pre">NumpyShared</span></code></a><ul>
<li><a class="reference internal" href="#PopPUNK.models.NumpyShared.dtype"><code class="docutils literal notranslate"><span class="pre">NumpyShared.dtype</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.NumpyShared.name"><code class="docutils literal notranslate"><span class="pre">NumpyShared.name</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.NumpyShared.shape"><code class="docutils literal notranslate"><span class="pre">NumpyShared.shape</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#PopPUNK.models.RefineFit"><code class="docutils literal notranslate"><span class="pre">RefineFit</span></code></a><ul>
<li><a class="reference internal" href="#PopPUNK.models.RefineFit.apply_threshold"><code class="docutils literal notranslate"><span class="pre">RefineFit.apply_threshold()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.RefineFit.assign"><code class="docutils literal notranslate"><span class="pre">RefineFit.assign()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.RefineFit.fit"><code class="docutils literal notranslate"><span class="pre">RefineFit.fit()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.RefineFit.load"><code class="docutils literal notranslate"><span class="pre">RefineFit.load()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.RefineFit.plot"><code class="docutils literal notranslate"><span class="pre">RefineFit.plot()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.RefineFit.save"><code class="docutils literal notranslate"><span class="pre">RefineFit.save()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#PopPUNK.models.assign_samples"><code class="docutils literal notranslate"><span class="pre">assign_samples()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.models.loadClusterFit"><code class="docutils literal notranslate"><span class="pre">loadClusterFit()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#network-py">network.py</a><ul>
<li><a class="reference internal" href="#PopPUNK.network.addQueryToNetwork"><code class="docutils literal notranslate"><span class="pre">addQueryToNetwork()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.add_self_loop"><code class="docutils literal notranslate"><span class="pre">add_self_loop()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.checkNetworkVertexCount"><code class="docutils literal notranslate"><span class="pre">checkNetworkVertexCount()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.cliquePrune"><code class="docutils literal notranslate"><span class="pre">cliquePrune()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.construct_dense_weighted_network"><code class="docutils literal notranslate"><span class="pre">construct_dense_weighted_network()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.construct_network_from_assignments"><code class="docutils literal notranslate"><span class="pre">construct_network_from_assignments()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.construct_network_from_df"><code class="docutils literal notranslate"><span class="pre">construct_network_from_df()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.construct_network_from_edge_list"><code class="docutils literal notranslate"><span class="pre">construct_network_from_edge_list()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.construct_network_from_sparse_matrix"><code class="docutils literal notranslate"><span class="pre">construct_network_from_sparse_matrix()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.cugraph_to_graph_tool"><code class="docutils literal notranslate"><span class="pre">cugraph_to_graph_tool()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.extractReferences"><code class="docutils literal notranslate"><span class="pre">extractReferences()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.fetchNetwork"><code class="docutils literal notranslate"><span class="pre">fetchNetwork()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.generate_minimum_spanning_tree"><code class="docutils literal notranslate"><span class="pre">generate_minimum_spanning_tree()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.getCliqueRefs"><code class="docutils literal notranslate"><span class="pre">getCliqueRefs()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.get_cugraph_triangles"><code class="docutils literal notranslate"><span class="pre">get_cugraph_triangles()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.get_vertex_list"><code class="docutils literal notranslate"><span class="pre">get_vertex_list()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.load_network_file"><code class="docutils literal notranslate"><span class="pre">load_network_file()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.networkSummary"><code class="docutils literal notranslate"><span class="pre">networkSummary()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.network_to_edges"><code class="docutils literal notranslate"><span class="pre">network_to_edges()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.printClusters"><code class="docutils literal notranslate"><span class="pre">printClusters()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.printExternalClusters"><code class="docutils literal notranslate"><span class="pre">printExternalClusters()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.print_network_summary"><code class="docutils literal notranslate"><span class="pre">print_network_summary()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.process_previous_network"><code class="docutils literal notranslate"><span class="pre">process_previous_network()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.process_weights"><code class="docutils literal notranslate"><span class="pre">process_weights()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.save_network"><code class="docutils literal notranslate"><span class="pre">save_network()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.sparse_mat_to_network"><code class="docutils literal notranslate"><span class="pre">sparse_mat_to_network()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.translate_network_indices"><code class="docutils literal notranslate"><span class="pre">translate_network_indices()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.vertex_betweenness"><code class="docutils literal notranslate"><span class="pre">vertex_betweenness()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.network.writeReferences"><code class="docutils literal notranslate"><span class="pre">writeReferences()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#refine-py">refine.py</a><ul>
<li><a class="reference internal" href="#PopPUNK.refine.NumpyShared"><code class="docutils literal notranslate"><span class="pre">NumpyShared</span></code></a><ul>
<li><a class="reference internal" href="#PopPUNK.refine.NumpyShared.dtype"><code class="docutils literal notranslate"><span class="pre">NumpyShared.dtype</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.refine.NumpyShared.name"><code class="docutils literal notranslate"><span class="pre">NumpyShared.name</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.refine.NumpyShared.shape"><code class="docutils literal notranslate"><span class="pre">NumpyShared.shape</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#PopPUNK.refine.check_search_range"><code class="docutils literal notranslate"><span class="pre">check_search_range()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.refine.expand_cugraph_network"><code class="docutils literal notranslate"><span class="pre">expand_cugraph_network()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.refine.growNetwork"><code class="docutils literal notranslate"><span class="pre">growNetwork()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.refine.likelihoodBoundary"><code class="docutils literal notranslate"><span class="pre">likelihoodBoundary()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.refine.multi_refine"><code class="docutils literal notranslate"><span class="pre">multi_refine()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.refine.newNetwork"><code class="docutils literal notranslate"><span class="pre">newNetwork()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.refine.newNetwork2D"><code class="docutils literal notranslate"><span class="pre">newNetwork2D()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.refine.readManualStart"><code class="docutils literal notranslate"><span class="pre">readManualStart()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.refine.refineFit"><code class="docutils literal notranslate"><span class="pre">refineFit()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-PopPUNK.plot">plot.py</a><ul>
<li><a class="reference internal" href="#PopPUNK.plot.createMicroreact"><code class="docutils literal notranslate"><span class="pre">createMicroreact()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.distHistogram"><code class="docutils literal notranslate"><span class="pre">distHistogram()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.drawMST"><code class="docutils literal notranslate"><span class="pre">drawMST()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.get_grid"><code class="docutils literal notranslate"><span class="pre">get_grid()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.outputsForCytoscape"><code class="docutils literal notranslate"><span class="pre">outputsForCytoscape()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.outputsForGrapetree"><code class="docutils literal notranslate"><span class="pre">outputsForGrapetree()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.outputsForMicroreact"><code class="docutils literal notranslate"><span class="pre">outputsForMicroreact()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.outputsForPhandango"><code class="docutils literal notranslate"><span class="pre">outputsForPhandango()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.plot_contours"><code class="docutils literal notranslate"><span class="pre">plot_contours()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.plot_dbscan_results"><code class="docutils literal notranslate"><span class="pre">plot_dbscan_results()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.plot_fit"><code class="docutils literal notranslate"><span class="pre">plot_fit()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.plot_refined_results"><code class="docutils literal notranslate"><span class="pre">plot_refined_results()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.plot_results"><code class="docutils literal notranslate"><span class="pre">plot_results()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.plot_scatter"><code class="docutils literal notranslate"><span class="pre">plot_scatter()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.plot.writeClusterCsv"><code class="docutils literal notranslate"><span class="pre">writeClusterCsv()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-PopPUNK.sparse_mst">sparse_mst.py</a></li>
<li><a class="reference internal" href="#module-PopPUNK.sketchlib">sketchlib.py</a><ul>
<li><a class="reference internal" href="#PopPUNK.sketchlib.addRandom"><code class="docutils literal notranslate"><span class="pre">addRandom()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.sketchlib.checkSketchlibLibrary"><code class="docutils literal notranslate"><span class="pre">checkSketchlibLibrary()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.sketchlib.checkSketchlibVersion"><code class="docutils literal notranslate"><span class="pre">checkSketchlibVersion()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.sketchlib.constructDatabase"><code class="docutils literal notranslate"><span class="pre">constructDatabase()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.sketchlib.createDatabaseDir"><code class="docutils literal notranslate"><span class="pre">createDatabaseDir()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.sketchlib.fitKmerCurve"><code class="docutils literal notranslate"><span class="pre">fitKmerCurve()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.sketchlib.getKmersFromReferenceDatabase"><code class="docutils literal notranslate"><span class="pre">getKmersFromReferenceDatabase()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.sketchlib.getSeqsInDb"><code class="docutils literal notranslate"><span class="pre">getSeqsInDb()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.sketchlib.getSketchSize"><code class="docutils literal notranslate"><span class="pre">getSketchSize()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.sketchlib.joinDBs"><code class="docutils literal notranslate"><span class="pre">joinDBs()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.sketchlib.queryDatabase"><code class="docutils literal notranslate"><span class="pre">queryDatabase()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.sketchlib.readDBParams"><code class="docutils literal notranslate"><span class="pre">readDBParams()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.sketchlib.removeFromDB"><code class="docutils literal notranslate"><span class="pre">removeFromDB()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-PopPUNK.utils">utils.py</a><ul>
<li><a class="reference internal" href="#PopPUNK.utils.check_and_set_gpu"><code class="docutils literal notranslate"><span class="pre">check_and_set_gpu()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.decisionBoundary"><code class="docutils literal notranslate"><span class="pre">decisionBoundary()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.get_match_search_depth"><code class="docutils literal notranslate"><span class="pre">get_match_search_depth()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.isolateNameToLabel"><code class="docutils literal notranslate"><span class="pre">isolateNameToLabel()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.iterDistRows"><code class="docutils literal notranslate"><span class="pre">iterDistRows()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.joinClusterDicts"><code class="docutils literal notranslate"><span class="pre">joinClusterDicts()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.listDistInts"><code class="docutils literal notranslate"><span class="pre">listDistInts()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.readIsolateTypeFromCsv"><code class="docutils literal notranslate"><span class="pre">readIsolateTypeFromCsv()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.readPickle"><code class="docutils literal notranslate"><span class="pre">readPickle()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.readRfile"><code class="docutils literal notranslate"><span class="pre">readRfile()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.read_rlist_from_distance_pickle"><code class="docutils literal notranslate"><span class="pre">read_rlist_from_distance_pickle()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.set_env"><code class="docutils literal notranslate"><span class="pre">set_env()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.setupDBFuncs"><code class="docutils literal notranslate"><span class="pre">setupDBFuncs()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.stderr_redirected"><code class="docutils literal notranslate"><span class="pre">stderr_redirected()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.storePickle"><code class="docutils literal notranslate"><span class="pre">storePickle()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.transformLine"><code class="docutils literal notranslate"><span class="pre">transformLine()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.utils.update_distance_matrices"><code class="docutils literal notranslate"><span class="pre">update_distance_matrices()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#visualise-py">visualise.py</a><ul>
<li><a class="reference internal" href="#PopPUNK.visualise.main"><code class="docutils literal notranslate"><span class="pre">main()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#web-py">web.py</a><ul>
<li><a class="reference internal" href="#PopPUNK.web.api"><code class="docutils literal notranslate"><span class="pre">api()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.web.calc_prevalence"><code class="docutils literal notranslate"><span class="pre">calc_prevalence()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.web.graphml_to_json"><code class="docutils literal notranslate"><span class="pre">graphml_to_json()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.web.highlight_cluster"><code class="docutils literal notranslate"><span class="pre">highlight_cluster()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.web.sketch_to_hdf5"><code class="docutils literal notranslate"><span class="pre">sketch_to_hdf5()</span></code></a></li>
<li><a class="reference internal" href="#PopPUNK.web.summarise_clusters"><code class="docutils literal notranslate"><span class="pre">summarise_clusters()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=1fc3131f"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/scripts/furo.js?v=32e29ea5"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>