<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Creating visualisations" href="visualisation.html" /><link rel="prev" title="Distributing PopPUNK models" href="model_distribution.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2023.09.10 -->
        <title>Query assignment (poppunk_assign) - PopPUNK 2.6.2 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">PopPUNK 2.6.2 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="_static/poppunk_v2.png" alt="Logo"/>
  </div>
  
  <span class="sidebar-brand-text">PopPUNK 2.6.2 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">PopPUNK documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="sketching.html">Sketching (<code class="docutils literal notranslate"><span class="pre">--create-db</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="qc.html">Data quality control (<code class="docutils literal notranslate"><span class="pre">--qc-db</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_fitting.html">Fitting new models (<code class="docutils literal notranslate"><span class="pre">--fit-model</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_distribution.html">Distributing PopPUNK models</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Query assignment (<code class="docutils literal notranslate"><span class="pre">poppunk_assign</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="visualisation.html">Creating visualisations</a></li>
<li class="toctree-l1"><a class="reference internal" href="mst.html">Minimum spanning trees</a></li>
<li class="toctree-l1"><a class="reference internal" href="subclustering.html">Subclustering with PopPIPE</a></li>
<li class="toctree-l1"><a class="reference internal" href="gpu.html">Using GPUs</a></li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Scripts</a></li>
<li class="toctree-l1"><a class="reference internal" href="poppunk_iterate.html">Iterative PopPUNK</a></li>
<li class="toctree-l1"><a class="reference internal" href="citing.html">Citing PopPUNK</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">Reference documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="roadmap.html">Roadmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="miscellaneous.html">Miscellaneous</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="query-assignment-poppunk-assign">
<h1>Query assignment (<code class="docutils literal notranslate"><span class="pre">poppunk_assign</span></code>)<a class="headerlink" href="#query-assignment-poppunk-assign" title="Permalink to this heading">#</a></h1>
<p>This is the recommended mode to use PopPUNK, as long as a database is available for
your species. If there is no database available, you can fit your own (<a class="reference internal" href="model_fitting.html"><span class="doc">Fitting new models (--fit-model)</span></a>).</p>
<p>Briefly, <a class="reference external" href="https://www.bacpop.org/poppunk/">download your reference database</a> and run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">poppunk_assign</span> <span class="o">--</span><span class="n">db</span> <span class="n">database</span> <span class="o">--</span><span class="n">query</span> <span class="n">qfile</span><span class="o">.</span><span class="n">txt</span> \
<span class="o">--</span><span class="n">output</span> <span class="n">poppunk_clusters</span> <span class="o">--</span><span class="n">threads</span> <span class="mi">8</span>
</pre></div>
</div>
<section id="downloading-a-database">
<h2>Downloading a database<a class="headerlink" href="#downloading-a-database" title="Permalink to this heading">#</a></h2>
<p>Current PopPUNK databases can be found here: <a class="reference external" href="https://www.bacpop.org/poppunk/">https://www.bacpop.org/poppunk/</a></p>
<p>We refer to sequences in the database as references, and those being added
as queries. The clusters assigned by PopPUNK are variable-length-k-mer clusters (VLKCs).</p>
<p>A database called <code class="docutils literal notranslate"><span class="pre">database</span></code> will contain the following files, in <code class="docutils literal notranslate"><span class="pre">database/</span></code>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">database.h5</span></code> – the sketches of the reference sequences generated by <code class="docutils literal notranslate"><span class="pre">pp-sketchlib</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">database.dists.npy</span></code> and <code class="docutils literal notranslate"><span class="pre">database.dists.pkl</span></code> – the core and accessory distances for
all pairwise comparisons in the sketch database.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">database_fit.npz</span></code> and <code class="docutils literal notranslate"><span class="pre">database_fit.pkl</span></code> – the model fit to the core and accessory distances.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">database_graph.gt</span></code> – the network defining the fit (loadable with <code class="docutils literal notranslate"><span class="pre">graph_tool</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">database_clusters.csv</span></code> – the PopPUNK clusters for the reference sequences.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">database.refs</span></code> – a minimal list of references needed to produce correct clusters.</p></li>
</ul>
<p>If the <code class="docutils literal notranslate"><span class="pre">.refs</span></code> file is missing, all of the samples in the sketch database will be
used in the distance calculations.</p>
<p>You can use the following arguments to individually target these items if necessary,
for example when using an alternative fit, or if split across different directories. The
examples below refer to the default database name:</p>
<ul class="simple">
<li><p>(required) <code class="docutils literal notranslate"><span class="pre">--db</span> <span class="pre">database</span></code> – the name of directory containing the .h5 file.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--distances</span> <span class="pre">database/database.dists</span></code> – prefix of the distances.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--model-dir</span> <span class="pre">database</span></code> – directory containing the model fit and network (dists + fit define the network).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--previous-clustering</span> <span class="pre">database</span></code> – directory containing the PopPUNK clusters for the references.</p></li>
</ul>
</section>
<section id="clustering-your-genomes">
<h2>Clustering your genomes<a class="headerlink" href="#clustering-your-genomes" title="Permalink to this heading">#</a></h2>
<p>Create a file which lists your sample names and paths to their sequence data. This file
has no header, is tab separated, and contains the sample name in the first column. Subsequent
columns may contain paths to either assembled or raw read data (the type will automatically
be inferred by checking for the presence of quality scores). Data may be gzipped or uncompressed:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">MS1</span> <span class="n">ms1_assembled</span><span class="o">.</span><span class="n">fa</span>
<span class="n">MS2</span> <span class="n">ms2_assembled</span><span class="o">.</span><span class="n">fa</span>
<span class="n">SM14</span>        <span class="n">SM14_1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">SM14_2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>Save this as <code class="docutils literal notranslate"><span class="pre">qfile.txt</span></code>. You’re now ready to cluster them!
Run the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">poppunk_assign</span> <span class="o">--</span><span class="n">db</span> <span class="n">database</span> <span class="o">--</span><span class="n">query</span> <span class="n">qfile</span><span class="o">.</span><span class="n">txt</span> \
<span class="o">--</span><span class="n">output</span> <span class="n">poppunk_clusters</span> <span class="o">--</span><span class="n">threads</span> <span class="mi">8</span>
</pre></div>
</div>
<p>This will first of all sketch your input genomes, saving them in <code class="docutils literal notranslate"><span class="pre">poppunk_clusters/poppunk_clusters.h5</span></code>.
If you need to rerun part of the analysis with different options this will automatically be picked up
and loaded.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="qc.html"><span class="doc">Data quality control (--qc-db)</span></a> does not apply to query sequences. A test for maximum accessory distance
will be made, but the program will only emit warnings and will run with all genomes
anyway. Most options for sketching will be taken from the reference database, but you
can still specify error filtering options from read input (<code class="docutils literal notranslate"><span class="pre">--min-kmer-count</span></code> and
<code class="docutils literal notranslate"><span class="pre">--exact-count</span></code>) and specify your input as <code class="docutils literal notranslate"><span class="pre">--strand-preserved</span></code>. See <a class="reference internal" href="sketching.html"><span class="doc">Sketching (--create-db)</span></a> for
more information on these options.</p>
</div>
<p>Next, core and accessory distances between your input sketches and those in the database
will be computed. This has complexity <span class="math notranslate nohighlight">\(O(RQ)\)</span> where <span class="math notranslate nohighlight">\(R\)</span> is the number of
samples in <code class="docutils literal notranslate"><span class="pre">database_references.refs</span></code> and <span class="math notranslate nohighlight">\(Q\)</span> is the number in <code class="docutils literal notranslate"><span class="pre">qfile.txt</span></code>. These distances
are then fed into the model and used to update the network, and therefore clusters.</p>
<p>The output will look something like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Graph</span><span class="o">-</span><span class="n">tools</span> <span class="n">OpenMP</span> <span class="n">parallelisation</span> <span class="n">enabled</span><span class="p">:</span> <span class="k">with</span> <span class="mi">4</span> <span class="n">threads</span>
<span class="n">PopPUNK</span> <span class="p">(</span><span class="n">POPulation</span> <span class="n">Partitioning</span> <span class="n">Using</span> <span class="n">Nucleotide</span> <span class="n">Kmers</span><span class="p">)</span>
        <span class="p">(</span><span class="k">with</span> <span class="n">backend</span><span class="p">:</span> <span class="n">sketchlib</span> <span class="n">v1</span><span class="mf">.5.1</span>
         <span class="n">sketchlib</span><span class="p">:</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">jlees</span><span class="o">/</span><span class="n">miniconda3</span><span class="o">/</span><span class="n">envs</span><span class="o">/</span><span class="n">pp</span><span class="o">-</span><span class="n">py38</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.8</span><span class="o">/</span><span class="n">site</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">pp_sketchlib</span><span class="o">.</span><span class="n">cpython</span><span class="o">-</span><span class="mi">38</span><span class="o">-</span><span class="n">darwin</span><span class="o">.</span><span class="n">so</span><span class="p">)</span>
<span class="n">Mode</span><span class="p">:</span> <span class="n">Assigning</span> <span class="n">clusters</span> <span class="n">of</span> <span class="n">query</span> <span class="n">sequences</span>

<span class="n">Sketching</span> <span class="n">genomes</span> <span class="n">using</span> <span class="mi">8</span> <span class="n">thread</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="n">Calculating</span> <span class="n">distances</span> <span class="n">using</span> <span class="mi">8</span> <span class="n">thread</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="n">Loading</span> <span class="n">previously</span> <span class="n">refined</span> <span class="n">model</span>
<span class="n">Network</span> <span class="n">loaded</span><span class="p">:</span> <span class="mi">2007</span> <span class="n">samples</span>
<span class="n">Found</span> <span class="n">novel</span> <span class="n">query</span> <span class="n">clusters</span><span class="o">.</span> <span class="n">Calculating</span> <span class="n">distances</span> <span class="n">between</span> <span class="n">them</span><span class="o">.</span>
<span class="n">Could</span> <span class="ow">not</span> <span class="n">find</span> <span class="n">random</span> <span class="n">match</span> <span class="n">chances</span> <span class="ow">in</span> <span class="n">database</span><span class="p">,</span> <span class="n">calculating</span> <span class="n">assuming</span> <span class="n">equal</span> <span class="n">base</span> <span class="n">frequencies</span>
<span class="n">Calculating</span> <span class="n">distances</span> <span class="n">using</span> <span class="mi">8</span> <span class="n">thread</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
</pre></div>
</div>
<p>Your VLKCs will be written to <code class="docutils literal notranslate"><span class="pre">poppunk_clusters/poppunk_clusters_clusters.csv</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Taxon</span><span class="p">,</span><span class="n">Cluster</span>
<span class="mi">21946_6_66</span><span class="p">,</span><span class="mi">9</span>
<span class="mi">22695_3_148</span><span class="p">,</span><span class="mi">9</span>
<span class="mi">22984_8_88</span><span class="p">,</span><span class="mi">9</span>
<span class="mi">21946_6_245</span><span class="p">,</span><span class="mi">116</span>
<span class="mi">21946_6_189</span><span class="p">,</span><span class="mi">814</span>
<span class="mi">22695_3_73</span><span class="p">,</span><span class="mi">814</span>
<span class="mi">21946_6_50</span><span class="p">,</span><span class="mi">422</span>
<span class="mi">21903_8_95</span><span class="p">,</span><span class="mi">148</span>
<span class="mi">21903_8_250</span><span class="p">,</span><span class="mi">301</span>
<span class="mi">22984_8_47</span><span class="p">,</span><span class="mi">70</span>
</pre></div>
</div>
<p>These names are identical to those used in the reference database, so retain
the same meaning between studies. If new clusters are found they will be numbered
in ascending order from largest to smallest, beginning from the end of the reference
clusters.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You may observe clusters merging (but never splitting). If your genomes
do cause clusters to merge this will be noted in the output, and the new
clusters will be named using the old ones. For example, if clusters 23 and 38
merged, the new cluster would be called 23_38.</p>
</div>
<p>By default, only the query genome clusters are included here. The reference genome
clusters are considered unchanged from the input. If there are many merges and you
wish to know their new cluster IDs, use <code class="docutils literal notranslate"><span class="pre">--update-db</span></code> (<a class="reference internal" href="#update-db"><span class="std std-ref">Updating the database</span></a>).</p>
<p>You can use <code class="docutils literal notranslate"><span class="pre">poppunk_visualise</span></code> to look at your results. Here’s an example output
to cytoscape, showing the clusters as colours, reference genomes as circles and
queries as triangles (open in a new tab to zoom on detail):</p>
<img alt="Network produced after query assignment" class="align-center" src="_images/assign_network.png" />
<section id="adding-external-cluster-labels-mlst-cc-etc">
<h3>Adding external cluster labels (MLST, CC etc)<a class="headerlink" href="#adding-external-cluster-labels-mlst-cc-etc" title="Permalink to this heading">#</a></h3>
<p>Add the <code class="docutils literal notranslate"><span class="pre">--external-clustering</span></code> argument to add a CSV file of cluster definitions
which the output will be additionally labelled with, and output to <code class="docutils literal notranslate"><span class="pre">database/database_external_clusters.csv</span></code>.
These can be any cluster definitions you wish, with as many columns as you like. A header row is required:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span><span class="p">,</span><span class="n">GPSC</span><span class="p">,</span><span class="n">MLST</span>
<span class="mi">23430_1_186</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">22</span>
<span class="mi">17794_6_29</span><span class="p">,</span><span class="mi">23</span><span class="p">,</span><span class="mi">43</span>
<span class="mi">12291_4_13</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span>
</pre></div>
</div>
<p>For each PopPUNK cluster, all the samples found in said cluster will be accumulated.
From these accumulated samples the external clusters will be collected, and assigned
to all of these examples. This may give you a one-to-one mapping between PopPUNK clusters
and your external cluster, or you may find multiple external clusters refer to the
PopPUNK cluster giving output such as <code class="docutils literal notranslate"><span class="pre">227;811;763;824</span></code>.</p>
</section>
<section id="using-a-model-fitted-with-indiv-refine">
<h3>Using a model fitted with <code class="docutils literal notranslate"><span class="pre">--indiv-refine</span></code><a class="headerlink" href="#using-a-model-fitted-with-indiv-refine" title="Permalink to this heading">#</a></h3>
<p>If the database was fitted with the refine fit mode, and <code class="docutils literal notranslate"><span class="pre">indiv-refine</span></code> you may have
a core distance boundary, accessory boundary and combined core-accessory boundary fit. The
default is to use the combined boundary, to use the others add <code class="docutils literal notranslate"><span class="pre">--core-only</span></code> or
<code class="docutils literal notranslate"><span class="pre">--accessory-only</span></code>.</p>
</section>
</section>
<section id="increasing-speed">
<h2>Increasing speed<a class="headerlink" href="#increasing-speed" title="Permalink to this heading">#</a></h2>
<p>Query assignment is the most efficient mode in which to run PopPUNK, typically requiring <span class="math notranslate nohighlight">\(Q\)</span> sketches and
<span class="math notranslate nohighlight">\(RQ\)</span> distances. If you are updating the database, this increases to <span class="math notranslate nohighlight">\(Q^2 + RQ\)</span>
distances. If you are assigning a very large number of queries you can run <code class="docutils literal notranslate"><span class="pre">poppunk_assign</span></code>
with <code class="docutils literal notranslate"><span class="pre">--update-db</span></code> repeatedly for batches of query input, as the <span class="math notranslate nohighlight">\(Q^2\)</span> term will
be reduced by clique-pruning at each iteration.</p>
<p>Straightforward ways to increase speed include:</p>
<ul class="simple">
<li><p>Add <code class="docutils literal notranslate"><span class="pre">--gpu-dist</span></code>, if you have a GPU available.</p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">--gpu-sketch</span></code>, if your input is all reads, and you have a GPU available. If
your input is a mix of assemblies and reads, run in two separate batches, with
the batch of reads using this option.</p></li>
<li><p>Increase <code class="docutils literal notranslate"><span class="pre">--threads</span></code>.</p></li>
</ul>
</section>
<section id="updating-the-database">
<span id="update-db"></span><h2>Updating the database<a class="headerlink" href="#updating-the-database" title="Permalink to this heading">#</a></h2>
<p>If you want to add your query genomes into the reference database so that they
can be used to inform future cluster assignment, this is as simple as adding the
<code class="docutils literal notranslate"><span class="pre">--update-db</span></code> option to the command above. This is particularly useful when novel
query clusters have been found – they will then be the consistent name for future assignments:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">poppunk_assign</span> <span class="o">--</span><span class="n">db</span> <span class="n">database</span> <span class="o">--</span><span class="n">query</span> <span class="n">qfile</span><span class="o">.</span><span class="n">txt</span> \
<span class="o">--</span><span class="n">output</span> <span class="n">poppunk_clusters</span> <span class="o">--</span><span class="n">threads</span> <span class="mi">8</span> <span class="o">--</span><span class="n">update</span><span class="o">-</span><span class="n">db</span>

<span class="n">Graph</span><span class="o">-</span><span class="n">tools</span> <span class="n">OpenMP</span> <span class="n">parallelisation</span> <span class="n">enabled</span><span class="p">:</span> <span class="k">with</span> <span class="mi">4</span> <span class="n">threads</span>
<span class="n">PopPUNK</span> <span class="p">(</span><span class="n">POPulation</span> <span class="n">Partitioning</span> <span class="n">Using</span> <span class="n">Nucleotide</span> <span class="n">Kmers</span><span class="p">)</span>
    <span class="p">(</span><span class="k">with</span> <span class="n">backend</span><span class="p">:</span> <span class="n">sketchlib</span> <span class="n">v1</span><span class="mf">.5.1</span>
    <span class="n">sketchlib</span><span class="p">:</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">jlees</span><span class="o">/</span><span class="n">miniconda3</span><span class="o">/</span><span class="n">envs</span><span class="o">/</span><span class="n">pp</span><span class="o">-</span><span class="n">py38</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.8</span><span class="o">/</span><span class="n">site</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">pp_sketchlib</span><span class="o">.</span><span class="n">cpython</span><span class="o">-</span><span class="mi">38</span><span class="o">-</span><span class="n">darwin</span><span class="o">.</span><span class="n">so</span><span class="p">)</span>
<span class="n">Mode</span><span class="p">:</span> <span class="n">Assigning</span> <span class="n">clusters</span> <span class="n">of</span> <span class="n">query</span> <span class="n">sequences</span>

<span class="n">Sketching</span> <span class="mi">28</span> <span class="n">genomes</span> <span class="n">using</span> <span class="mi">4</span> <span class="n">thread</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="n">Writing</span> <span class="n">sketches</span> <span class="n">to</span> <span class="n">file</span>
<span class="n">Calculating</span> <span class="n">distances</span> <span class="n">using</span> <span class="mi">4</span> <span class="n">thread</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="n">Loading</span> <span class="n">BGMM</span> <span class="mi">2</span><span class="n">D</span> <span class="n">Gaussian</span> <span class="n">model</span>
<span class="n">Network</span> <span class="n">loaded</span><span class="p">:</span> <span class="mi">18</span> <span class="n">samples</span>
<span class="n">Calculating</span> <span class="nb">all</span> <span class="n">query</span><span class="o">-</span><span class="n">query</span> <span class="n">distances</span>
<span class="n">Could</span> <span class="ow">not</span> <span class="n">find</span> <span class="n">random</span> <span class="n">match</span> <span class="n">chances</span> <span class="ow">in</span> <span class="n">database</span><span class="p">,</span> <span class="n">calculating</span> <span class="n">assuming</span> <span class="n">equal</span> <span class="n">base</span> <span class="n">frequencies</span>
<span class="n">Calculating</span> <span class="n">distances</span> <span class="n">using</span> <span class="mi">4</span> <span class="n">thread</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="n">Updating</span> <span class="n">reference</span> <span class="n">database</span> <span class="n">to</span> <span class="n">poppunk_clusters</span>
<span class="n">Removing</span> <span class="mi">27</span> <span class="n">sequences</span>

<span class="n">Done</span>
</pre></div>
</div>
<p>The new database contains all of the reference sequences, and all of your query sequences.
The <code class="docutils literal notranslate"><span class="pre">poppunk_clusters</span></code> folder will now contain all of the files of a reference
database listed above, except for the model. You can use <code class="docutils literal notranslate"><span class="pre">--model-dir</span></code> to target
this for future assignment, or copy it over yourself. Alternatively, if you run
with the same <code class="docutils literal notranslate"><span class="pre">--output</span></code> folder as <code class="docutils literal notranslate"><span class="pre">--ref-db</span></code>, adding <code class="docutils literal notranslate"><span class="pre">--overwrite</span></code>, the original
input folder will contain the updated database containing everything needed.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This mode can take longer to run with large numbers of input query genomes,
as it will calculate all <span class="math notranslate nohighlight">\(Q^2\)</span> query-query distances, rather than
just those found in novel query clusters.</p>
</div>
</section>
<section id="visualising-results">
<h2>Visualising results<a class="headerlink" href="#visualising-results" title="Permalink to this heading">#</a></h2>
<p>If you wish to produce visualisations from query assignment results the best
way to do this is to run with <code class="docutils literal notranslate"><span class="pre">--update-db</span></code>, and then run <code class="docutils literal notranslate"><span class="pre">poppunk_visualise</span></code>
on the output directory, as if visualising a full reference fit.</p>
<p>However, it is possible to run directly on the outputs by adding a <code class="docutils literal notranslate"><span class="pre">--ref-db</span></code>
as used in the assign command, and a <code class="docutils literal notranslate"><span class="pre">--query-db</span></code> which points to the <code class="docutils literal notranslate"><span class="pre">--output</span></code>
directory used in the assign command. In this mode isolates will be annotated
depending on whether they were a query or reference input.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Without <code class="docutils literal notranslate"><span class="pre">--update-db</span></code>, visualisation is required to recalculate all query-query distances
each time it is called. If your query set is large and you want repeated visualisations,
run <code class="docutils literal notranslate"><span class="pre">poppunk_assign</span></code> with <code class="docutils literal notranslate"><span class="pre">--update-db</span></code>.</p>
</div>
<p>See <a class="reference internal" href="visualisation.html"><span class="doc">Creating visualisations</span></a> for more details.</p>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="visualisation.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Creating visualisations</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="model_distribution.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Distributing PopPUNK models</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2018-2023, John Lees and Nicholas Croucher
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Query assignment (<code class="docutils literal notranslate"><span class="pre">poppunk_assign</span></code>)</a><ul>
<li><a class="reference internal" href="#downloading-a-database">Downloading a database</a></li>
<li><a class="reference internal" href="#clustering-your-genomes">Clustering your genomes</a><ul>
<li><a class="reference internal" href="#adding-external-cluster-labels-mlst-cc-etc">Adding external cluster labels (MLST, CC etc)</a></li>
<li><a class="reference internal" href="#using-a-model-fitted-with-indiv-refine">Using a model fitted with <code class="docutils literal notranslate"><span class="pre">--indiv-refine</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#increasing-speed">Increasing speed</a></li>
<li><a class="reference internal" href="#updating-the-database">Updating the database</a></li>
<li><a class="reference internal" href="#visualising-results">Visualising results</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=1fc3131f"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/scripts/furo.js?v=32e29ea5"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>