<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Minimum spanning trees" href="mst.html" /><link rel="prev" title="Query assignment (poppunk_assign)" href="query_assignment.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2023.09.10 -->
        <title>Creating visualisations - PopPUNK 2.6.2 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">PopPUNK 2.6.2 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="_static/poppunk_v2.png" alt="Logo"/>
  </div>
  
  <span class="sidebar-brand-text">PopPUNK 2.6.2 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">PopPUNK documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="sketching.html">Sketching (<code class="docutils literal notranslate"><span class="pre">--create-db</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="qc.html">Data quality control (<code class="docutils literal notranslate"><span class="pre">--qc-db</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_fitting.html">Fitting new models (<code class="docutils literal notranslate"><span class="pre">--fit-model</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_distribution.html">Distributing PopPUNK models</a></li>
<li class="toctree-l1"><a class="reference internal" href="query_assignment.html">Query assignment (<code class="docutils literal notranslate"><span class="pre">poppunk_assign</span></code>)</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Creating visualisations</a></li>
<li class="toctree-l1"><a class="reference internal" href="mst.html">Minimum spanning trees</a></li>
<li class="toctree-l1"><a class="reference internal" href="subclustering.html">Subclustering with PopPIPE</a></li>
<li class="toctree-l1"><a class="reference internal" href="gpu.html">Using GPUs</a></li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Scripts</a></li>
<li class="toctree-l1"><a class="reference internal" href="poppunk_iterate.html">Iterative PopPUNK</a></li>
<li class="toctree-l1"><a class="reference internal" href="citing.html">Citing PopPUNK</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">Reference documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="roadmap.html">Roadmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="miscellaneous.html">Miscellaneous</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="creating-visualisations">
<h1>Creating visualisations<a class="headerlink" href="#creating-visualisations" title="Permalink to this heading">#</a></h1>
<p>We have moved visualisation tools into their own program <code class="docutils literal notranslate"><span class="pre">poppunk_visualise</span></code>, both
to reinforce our commitment to UK spellings, and so that you can rerun visualisations
with different outputs and settings without rerunning the other parts of the code.</p>
<p>Starting with either a full database where you have fitted a model (<a class="reference internal" href="model_fitting.html"><span class="doc">Fitting new models (--fit-model)</span></a>), or
output where you have assigned queries using an existing database (<a class="reference internal" href="query_assignment.html"><span class="doc">Query assignment (poppunk_assign)</span></a>), you
can create outputs for external interactive visualisation tools:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://microreact.org/">Microreact</a> – a genomic epidemiology visualisation tool, displaying clusters, phylogeny and accessory clustering.</p></li>
<li><p><a class="reference external" href="https://achtman-lab.github.io/GrapeTree/MSTree_holder.html">GrapeTree</a> – a tool to visualise strains (designed for cgMLST).</p></li>
<li><p><a class="reference external" href="https://jameshadfield.github.io/phandango/#/">Phandango</a> – visualisation linking genomes and phylogenies.</p></li>
<li><p><a class="reference external" href="https://cytoscape.org/">Cytoscape</a> – a network visualisation tool, which can be used to create, view and manipulate a layout of the graph.</p></li>
</ul>
<p>At least one of these output types must be specified as a flag.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>If you run a visualisation on output from query assignment (<a class="reference internal" href="query_assignment.html"><span class="doc">Query assignment (poppunk_assign)</span></a>)
this will not contain all the necessary distances, and they will be calculated before
the visualisation files are produced.
You will see a message <code class="docutils literal notranslate"><span class="pre">Note:</span> <span class="pre">Distance</span> <span class="pre">will</span> <span class="pre">be</span> <span class="pre">extended</span> <span class="pre">to</span> <span class="pre">full</span> <span class="pre">all-vs-all</span> <span class="pre">distances</span></code>.
If you are running multiple visualisations this calculation will be completed every time. To avoid
this re-run your assignment with <code class="docutils literal notranslate"><span class="pre">--update-db</span></code>, which will add these distances in permanently.</p>
</div>
<section id="common-options">
<h2>Common options<a class="headerlink" href="#common-options" title="Permalink to this heading">#</a></h2>
<p>Some typical commands for various input settings (with <code class="docutils literal notranslate"><span class="pre">--microreact</span></code>, but this can
be altered to any output type) with a database <code class="docutils literal notranslate"><span class="pre">example</span></code>.</p>
<p>Visualisation of a full database:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">poppunk_visualise</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">db</span> <span class="n">example_db</span> <span class="o">--</span><span class="n">output</span> <span class="n">example_viz</span> <span class="o">--</span><span class="n">microreact</span>
</pre></div>
</div>
<p>Visualisation after query assignment:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">poppunk_visualise</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">db</span> <span class="n">example_db</span> <span class="o">--</span><span class="n">query</span><span class="o">-</span><span class="n">db</span> <span class="n">example_query</span> <span class="o">--</span><span class="n">output</span> <span class="n">example_viz</span> <span class="o">--</span><span class="n">microreact</span>
</pre></div>
</div>
<p>Visualisation when sketches and models are in different folders:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">poppunk_visualise</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">db</span> <span class="n">example_db</span> <span class="o">--</span><span class="n">previous</span><span class="o">-</span><span class="n">clustering</span> <span class="n">example_lineages</span><span class="o">/</span><span class="n">example_lineages_lineages</span><span class="o">.</span><span class="n">csv</span> \
 <span class="o">--</span><span class="n">model</span><span class="o">-</span><span class="nb">dir</span> <span class="n">example_lineages</span> <span class="o">--</span><span class="n">output</span> <span class="n">example_viz</span> <span class="o">--</span><span class="n">microreact</span>
</pre></div>
</div>
<p>Visualisation with a lineage model, which has been queried (query-query distances must be provided):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">poppunk_visualise</span> <span class="o">--</span><span class="n">distances</span> <span class="n">example_query</span><span class="o">/</span><span class="n">example_query</span><span class="o">.</span><span class="n">dists</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">db</span> <span class="n">example_db</span> \
 <span class="o">--</span><span class="n">model</span><span class="o">-</span><span class="nb">dir</span> <span class="n">example_lineages</span> <span class="o">--</span><span class="n">query</span><span class="o">-</span><span class="n">db</span> <span class="n">example_lineage_query</span> \
 <span class="o">--</span><span class="n">output</span> <span class="n">example_viz</span> <span class="o">--</span><span class="n">microreact</span>
</pre></div>
</div>
<p>Notable modifiers include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--include-files</span></code> – give a file with a subset of names to be included in the visualisation.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--external-clustering</span></code> – other cluster names to map to strains (such as MLST, serotype etc),
as described in model fitting and query assignment.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--info-csv</span></code> – similar to the above, but a CSV which is simply (inner-)joined to the output on sample name.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--rapidnj</span></code> – the location of a <a class="reference external" href="https://birc.au.dk/software/rapidnj/">rapidnj</a> binary,
used to build the core NJ tree. We highly recommend using this for any tree-building (and is included with
the conda install). This defaults to <code class="docutils literal notranslate"><span class="pre">rapidnj</span></code> on the PATH. Set blank to use dendropy instead (slower, especially
for large datasets).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--core-only</span></code>/<code class="docutils literal notranslate"><span class="pre">--accessory-only</span></code> – use the core or accessory fit from an individually refined model (see <a class="reference internal" href="model_fitting.html#indiv-refine"><span class="std std-ref">Using core/accessory only</span></a>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--threads</span></code>, <code class="docutils literal notranslate"><span class="pre">--gpu-dist</span></code>, <code class="docutils literal notranslate"><span class="pre">--deviceid</span></code>, <code class="docutils literal notranslate"><span class="pre">--strand-preserved</span></code> – querying options used if extra distance calculations are needed.
To avoid these, rerun your query with <code class="docutils literal notranslate"><span class="pre">--update-db</span></code>.</p></li>
</ul>
</section>
<section id="microreact">
<h2>Microreact<a class="headerlink" href="#microreact" title="Permalink to this heading">#</a></h2>
<p>Adding the <code class="docutils literal notranslate"><span class="pre">--microreact</span></code> flag will create the following files:</p>
<ul class="simple">
<li><p>_microreact_clusters.csv – the strain or lineage assignments with headers formatted for microreact, plus anything from <code class="docutils literal notranslate"><span class="pre">--info-csv</span></code>.</p></li>
<li><p>_core_NJ.nwk – a neighbour-joining tree from the core distances.</p></li>
<li><p>_perplexity20.0_accessory_mandrake.dot – a 2D embedding of the accessory distances, produced using <a class="reference external" href="https://github.com/bacpop/mandrake">mandrake</a> (in this case with
<a class="reference internal" href="#perplexity"><span class="std std-ref">Setting the perplexity parameter for mandrake</span></a> 20).</p></li>
</ul>
<p>From version 2.5.0 this will also include:</p>
<ul class="simple">
<li><dl class="simple">
<dt>.microreact – a Microreact compatible JSON containing all of the above, which</dt><dd><p>can be uploaded directly.</p>
</dd>
</dl>
</li>
</ul>
<p>If you add <code class="docutils literal notranslate"><span class="pre">--api-key</span></code> and provide <cite>your account’s API key &lt;https://docs.microreact.org/api/access-tokens&gt;__</cite>
this will automatically create an instance, and the URL will be output to the terminal.</p>
<p>Otherwise, open <a class="reference external" href="https://microreact.org/upload">https://microreact.org/upload</a> in your browser, and drag and drop these three files
to create your visualisation. Here is the result of running the visualisation on the
<em>Listeria</em> BGMM model:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">poppunk_visualise</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">db</span> <span class="n">listeria</span> <span class="o">--</span><span class="n">microreact</span> <span class="o">--</span><span class="n">threads</span> <span class="mi">8</span>

<span class="n">Graph</span><span class="o">-</span><span class="n">tools</span> <span class="n">OpenMP</span> <span class="n">parallelisation</span> <span class="n">enabled</span><span class="p">:</span> <span class="k">with</span> <span class="mi">8</span> <span class="n">threads</span>
<span class="n">PopPUNK</span><span class="p">:</span> <span class="n">visualise</span>
<span class="n">Loading</span> <span class="n">BGMM</span> <span class="mi">2</span><span class="n">D</span> <span class="n">Gaussian</span> <span class="n">model</span>
<span class="n">Completed</span> <span class="n">model</span> <span class="n">loading</span>
<span class="n">Building</span> <span class="n">phylogeny</span>
<span class="n">Writing</span> <span class="n">microreact</span> <span class="n">output</span>
<span class="n">Parsed</span> <span class="n">data</span><span class="p">,</span> <span class="n">now</span> <span class="n">writing</span> <span class="n">to</span> <span class="n">CSV</span>
<span class="n">Running</span> <span class="n">mandrake</span>
<span class="n">Running</span> <span class="n">on</span> <span class="n">CPU</span>
<span class="n">Preprocessing</span> <span class="mi">128</span> <span class="n">samples</span> <span class="k">with</span> <span class="n">perplexity</span> <span class="o">=</span> <span class="mi">20</span> <span class="n">took</span> <span class="mi">0</span><span class="n">ms</span>
<span class="n">Optimizing</span>   <span class="n">Progress</span><span class="p">:</span> <span class="mf">99.9</span><span class="o">%</span><span class="p">,</span> <span class="n">eta</span><span class="o">=</span><span class="mf">0.0010</span><span class="p">,</span> <span class="n">Eq</span><span class="o">=</span><span class="mf">0.2583546852</span><span class="p">,</span> <span class="n">clashes</span><span class="o">=</span><span class="mf">2.1</span><span class="o">%</span>
<span class="n">Optimizing</span> <span class="n">done</span> <span class="ow">in</span> <span class="mi">30</span><span class="n">s</span>
<span class="n">Provide</span> <span class="o">--</span><span class="n">api</span><span class="o">-</span><span class="n">key</span> <span class="n">to</span> <span class="n">create</span> <span class="n">microreact</span> <span class="n">automatically</span>

<span class="n">Done</span>
</pre></div>
</div>
<p>This can be viewed at <a class="reference external" href="https://microreact.org/project/3JAZKqzJiaNyViWXindNLv-listeria-poppunk-perplexity-20">https://microreact.org/project/3JAZKqzJiaNyViWXindNLv-listeria-poppunk-perplexity-20</a>:</p>
<img alt="Microreact page for Listeria monocytogenes" class="align-center" src="_images/microreact.png" />
<p>Useful controls include the tree shape, accessed with the control slider in the
top right of the phylogeny page, and the metadata labels, accessed with the ‘eye’
on the right of the page. When visualising lineages, changing the ‘Colour by’ is useful
to compare results from different ranks.</p>
<section id="setting-the-perplexity-parameter-for-mandrake">
<span id="perplexity"></span><h3>Setting the perplexity parameter for mandrake<a class="headerlink" href="#setting-the-perplexity-parameter-for-mandrake" title="Permalink to this heading">#</a></h3>
<p>In mandrake an embedding of the accessory genome distances is found which
represents local structure of the data. Isolates with similar accessory content
will visually appear in clusters together.</p>
<p>The perplexity sets a guess about the number of close neighbours each point
has, and is a trade-off between local and global structure. t-SNE (and by extension mandrake) is reasonably
robust to changes in the perplexity parameter (set with <code class="docutils literal notranslate"><span class="pre">--perplexity</span></code> when
creating microreact output with <code class="docutils literal notranslate"><span class="pre">--microreact</span></code>),
however we would recommend trying a few values to get
a good embedding for the accessory distances.</p>
<p>There is a good discussion of the effect of perplexity <a class="reference external" href="https://distill.pub/2016/misread-tsne/">here</a>
and the sklearn documentation shows some examples of the effect of <a class="reference external" href="http://scikit-learn.org/stable/auto_examples/manifold/plot_t_sne_perplexity.html">changing
perplexity</a>.
In mandrake, points will usually appear ‘tighter’ than in t-SNE, and form more obvious clusters.</p>
<p>In the example with <em>Listeria monocytogenes</em> above, a perplexity of 20 gives clear clustering of
the accessory genome content, condordant with the core genome structure (<a class="reference external" href="https://microreact.org/project/3JAZKqzJiaNyViWXindNLv-listeria-poppunk-perplexity-20">data</a>):</p>
<p>With a lower perplexity of 5, the clustering is not as tight, but it still looks ok
(<a class="reference external" href="https://microreact.org/project/tXHmDR4NRfmTeemfjqbzip-listeria-poppunk-perplexity-5">data</a>):</p>
<img alt="Microreact plot of results with perplexity = 5" class="align-center" src="_images/microreact_perplexity5.png" />
<p>30 is a good default, but you may wish to try other values, particularly with
larger or smaller datasets. You can re-run the mandrake using the <code class="docutils literal notranslate"><span class="pre">poppunk_mandrake</span></code>
command, providing the distances from the previous run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">poppunk_mandrake</span> <span class="o">--</span><span class="n">distances</span> <span class="n">strain_db</span><span class="o">/</span><span class="n">strain_db</span><span class="o">.</span><span class="n">dists</span> <span class="o">--</span><span class="n">output</span> <span class="n">strain_db</span> \
<span class="o">--</span><span class="n">perplexity</span> <span class="mi">50</span>
</pre></div>
</div>
</section>
</section>
<section id="grapetree">
<h2>GrapeTree<a class="headerlink" href="#grapetree" title="Permalink to this heading">#</a></h2>
<p>Adding the <code class="docutils literal notranslate"><span class="pre">--grapetree</span></code> flag will create:</p>
<ul class="simple">
<li><p>_microreact_clusters.csv – the strain or lineage assignments with headers formatted for grapetree, plus anything from <code class="docutils literal notranslate"><span class="pre">--info-csv</span></code>.</p></li>
<li><p>_core_NJ.nwk – a neighbour-joining tree from the core distances.</p></li>
</ul>
<p>Open <a class="reference external" href="https://achtman-lab.github.io/GrapeTree/MSTree_holder.html">https://achtman-lab.github.io/GrapeTree/MSTree_holder.html</a> in your browser, and use
the ‘Load files’ button once for each of the files to add the tree and strain assignments to
GrapeTree. This will display an unrooted tree with your clusters:</p>
<img alt="Grapetree visualisation of results" class="align-center" src="_images/grapetree.png" />
<p>One of GrapeTree’s key features is the ability to collapse branches, and condense information
into nodes. By going to Tree Layout -&gt; Branch style -&gt; Collapse branches, and setting the long
branch to be shortened, one can obtain a view which shows strain prevalence and relationships:</p>
<img alt="Grapetree visualisation of results" class="align-center" src="_images/grapetree_collapse.png" />
<p>There is also a handy ‘Export to Microreact’ button in GrapeTree, though this will
not include the accessory embedding, so you may wish to add the <code class="docutils literal notranslate"><span class="pre">--microreact</span></code> flag
and generate the files yourself.</p>
</section>
<section id="phandango">
<h2>Phandango<a class="headerlink" href="#phandango" title="Permalink to this heading">#</a></h2>
<p>Adding the <code class="docutils literal notranslate"><span class="pre">--phandango</span></code> flag will create:</p>
<ul class="simple">
<li><p>_phandango_clusters.csv – the strain or lineage assignments with headers formatted for phandango, plus anything from <code class="docutils literal notranslate"><span class="pre">--info-csv</span></code>.</p></li>
<li><p>_core_NJ.tree – a neighbour-joining tree from the core distances.</p></li>
</ul>
<p>Open <a class="reference external" href="https://www.phandango.net">https://www.phandango.net</a> in your browser, and use
the ‘Load files’ button once for each of the files to add the tree and strain assignments to
GrapeTree. This will display the tree with your clusters:</p>
<img alt="Phandango visualisation of results" class="align-center" src="_images/phandango.png" />
<p>Press ‘s’ to access settings, and ‘p’ to create an .svg file. Phandango is most useful
with a genome (.gff file), and either a plot of recombination, accessory genome analysis
or GWAS results. See the documentation for more information.</p>
</section>
<section id="cytoscape">
<span id="cytoscape-view"></span><h2>Cytoscape<a class="headerlink" href="#cytoscape" title="Permalink to this heading">#</a></h2>
<p>Cytoscape is different from the above modes as it creates a layout and visualisation of
the graph used to create strains from distances. This can be useful for more detailed
investigation of network scores, particularly in strains which have less than perfect transitivity.</p>
<p>Add the <code class="docutils literal notranslate"><span class="pre">--cytoscape</span></code> option, and also <code class="docutils literal notranslate"><span class="pre">--network-file</span></code> to point to the
network you wish to visualise:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">poppunk_visualise</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">db</span> <span class="n">listeria</span> <span class="o">--</span><span class="n">cytoscape</span> <span class="o">--</span><span class="n">network</span><span class="o">-</span><span class="n">file</span> <span class="n">listeria</span><span class="o">/</span><span class="n">listeria_graph</span><span class="o">.</span><span class="n">gt</span>

<span class="n">Graph</span><span class="o">-</span><span class="n">tools</span> <span class="n">OpenMP</span> <span class="n">parallelisation</span> <span class="n">enabled</span><span class="p">:</span> <span class="k">with</span> <span class="mi">1</span> <span class="n">threads</span>
<span class="n">PopPUNK</span><span class="p">:</span> <span class="n">visualise</span>
<span class="n">Loading</span> <span class="n">BGMM</span> <span class="mi">2</span><span class="n">D</span> <span class="n">Gaussian</span> <span class="n">model</span>
<span class="n">Writing</span> <span class="n">cytoscape</span> <span class="n">output</span>
<span class="n">Network</span> <span class="n">loaded</span><span class="p">:</span> <span class="mi">128</span> <span class="n">samples</span>
<span class="n">Parsed</span> <span class="n">data</span><span class="p">,</span> <span class="n">now</span> <span class="n">writing</span> <span class="n">to</span> <span class="n">CSV</span>

<span class="n">Done</span>
</pre></div>
</div>
<p>Which will create:</p>
<ul class="simple">
<li><p>_cytoscape.csv – the strain or lineage assignments with headers formatted for cytoscape, plus anything from <code class="docutils literal notranslate"><span class="pre">--info-csv</span></code>.</p></li>
<li><p>_cytoscape.graphml – the network in graphml format.</p></li>
</ul>
<p>The .graphml file is an XML file which contains definitions of the nodes (samples)
and edges (within-strain distances) connecting them. If you used <code class="docutils literal notranslate"><span class="pre">--graph-weights</span></code>
when you fitted your model the edges will be annotated with their Euclidean distances
in the ‘weight’ attribute (which you will need to tell cytoscape). These can be added
with the <code class="docutils literal notranslate"><span class="pre">poppunk_add_weights</span></code> script if this flag was not used.</p>
<p>Open <a class="reference external" href="http://www.cytoscape.org/">cytoscape</a> and drag and drop the .graphml
file onto the window to import the network. Import -&gt; table -&gt; file to load the
CSV. Click ‘Select None’ then add the ‘id’ column as a key, and any required
metadata columns (at least the ‘Cluster’ column) as attributes. Make sure
‘Node Table Columns’ is selected as the data type.</p>
<p>The graphml file does not contain a layout for the graph, that is, positions of
nodes and edges are not specified for a visualisation. These will be calculated by cytoscape,
automatically for small graphs, and with the ‘Layout’ menu for larger graphs. The ‘Prefuse force directed layout’
or ‘yFiles Organic Layout’ work well. Select the ‘weight’ dropdown to use the edge-lengths
when drawing the network.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>We have found that making graphs with &gt;10k nodes may exceed the memory on a typical
laptop. To view larger graphs, first splitting into subgraphs of each connected component
is very helpful. Older versions of cytoscape allowed you to split the graph into connected
components, but newer versions have removed this feature. This can be done programmatically
with <code class="docutils literal notranslate"><span class="pre">networkx</span></code> or <code class="docutils literal notranslate"><span class="pre">graph-tool</span></code> in python, or <code class="docutils literal notranslate"><span class="pre">igraph</span></code> in R.</p>
</div>
<p>Click on ‘Style’ and change the node fill colour to be by cluster, the mapping
type as discrete, then right click to autogenerate a colour scheme (‘Random’ is usually best). You can
also modify the node size and shape here. Here is the <em>Listeria</em> example, using edge weights in the layout:</p>
<img alt="Cytoscape plot of network" class="align-center" src="_images/cytoscape.png" />
<p>If you used assign query mode you will also have a column with ‘Query’ or ‘Reference’, which can
be used to map to different shapes or colours:</p>
<img alt="Network produced after query assignment" class="align-center" src="_images/assign_network.png" />
<p>Adding an info CSV, or loading external tables directly into cytoscapes gives further options
for investigating individual strains:</p>
<img alt="Network with added annotation" class="align-center" src="_images/cytoscape_gpsc.png" />
<p>In some cases, edges which are between strain links may have been erroneously included
in the network. This could be due to poor model fit, or a poor quality
sequence. Use Tools -&gt; NetworkAnalyzer -&gt; Analyze Network to compute
information for each node and edge. It may help to analyze connected components separately.
They can be split under Tools -&gt; NetworkAnalyzer -&gt; Subnetwork Creation.</p>
<p>Here is an example where an errant node is connecting two clusters into one
large cluster, which should be split:</p>
<img alt="Cytoscape plot of network" class="align-center" src="_images/cytoscape_component.png" />
<p>The incorrect node in question has a low CluteringCoefficient and high Stress.
The EdgeBetweeness of its connections are also high. Sorting the node and edge
tables by these columns can find individual problems such as this.</p>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="mst.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Minimum spanning trees</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="query_assignment.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Query assignment (<code class="docutils literal notranslate"><span class="pre">poppunk_assign</span></code>)</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2018-2023, John Lees and Nicholas Croucher
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Creating visualisations</a><ul>
<li><a class="reference internal" href="#common-options">Common options</a></li>
<li><a class="reference internal" href="#microreact">Microreact</a><ul>
<li><a class="reference internal" href="#setting-the-perplexity-parameter-for-mandrake">Setting the perplexity parameter for mandrake</a></li>
</ul>
</li>
<li><a class="reference internal" href="#grapetree">GrapeTree</a></li>
<li><a class="reference internal" href="#phandango">Phandango</a></li>
<li><a class="reference internal" href="#cytoscape">Cytoscape</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=1fc3131f"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/scripts/furo.js?v=32e29ea5"></script>
    </body>
</html>